// @splendidlabz/utils/dist/cjs/lib/sse.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sse_exports = {};
__export(sse_exports, {
  createSSE: () => createSSE,
  parseSSE: () => parseSSE
});
module.exports = __toCommonJS(sse_exports);
var import_json = require("./objects/json.js");
var import_omit_empty = require("./objects/omit-empty.js");
function createSSE({ event, data, id, retry }) {
  const eventString = event ? `event: ${event}
` : "";
  const idString = id ? `id: ${id}
` : "";
  const retryString = retry ? `retry: ${retry}
` : "";
  if (typeof data === "object") data = JSON.stringify(data);
  const dataString = data ? `data: ${data}
` : "";
  return `${eventString}${dataString}${idString}${retryString}
`;
}
function parseSSE(message) {
  const lines = message.split("\n");
  const result = {
    event: "message",
    // Default event type
    data: "",
    id: null,
    retry: null
  };
  for (const line of lines) {
    if (line.startsWith("data:")) {
      const data = line.slice(5).trim();
      result.data = (0, import_json.parseJSON)(data);
      continue;
    }
    if (line.startsWith("event:")) {
      result.event = line.slice(6).trim();
      continue;
    }
    if (line.startsWith("id:")) {
      result.id = line.slice(3).trim();
      continue;
    }
    if (line.startsWith("retry:")) {
      result.retry = parseInt(line.slice(6).trim(), 10);
      continue;
    }
  }
  const ret = (0, import_omit_empty.omitEmpty)(result);
  return ret.data ? ret : null;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createSSE,
  parseSSE
});
