// @splendidlabz/utils/src/lib/objects/mix/mix.js
const DEFAULT_OPTIONS = { array: 'replace' } // Can be concat or replace

/**
 * Deep merges multiple objects into a new object with intelligent handling of arrays and nested objects.
 *
 * @param {...Object} sources - Objects to merge. Null and undefined values are skipped.
 * @return {Object} A new object containing the merged properties from all sources.
 *
 * @example
 * const obj1 = { a: 1, b: { x: 10 } }
 * const obj2 = { b: { y: 20 }, c: 3 }
 * const result = mix(obj1, obj2)
 * // Returns: { a: 1, b: { x: 10, y: 20 }, c: 3 }
 */
export const mix = createMix()
export const concatMix = createMix({ array: 'concat' })

/**
 * Creates a customized mix function with specific options for array handling.
 *
 * @param {Object} [userOptions={}] - Configuration options for the mix function.
 * @param {('replace'|'concat')} [userOptions.array='replace'] - How to handle array merging:
 *   - 'replace': Arrays from later sources completely replace arrays from earlier sources
 *   - 'concat': Arrays are concatenated together
 * @return {Function} A mix function configured with the provided options.
 *
 * @example
 * // Create mix function that concatenates arrays
 * const concatMix = createMix({ array: 'concat' })
 * const obj1 = { items: [1, 2] }
 * const obj2 = { items: [3, 4] }
 * const result = concatMix(obj1, obj2)
 * // Returns: { items: [1, 2, 3, 4] }
 *
 * @example
 * // Create mix function that replaces arrays (default behavior)
 * const replaceMix = createMix({ array: 'replace' })
 * const obj1 = { items: [1, 2] }
 * const obj2 = { items: [3, 4] }
 * const result = replaceMix(obj1, obj2)
 * // Returns: { items: [3, 4] }
 */
export function createMix(userOptions = {}) {
  const options = { ...DEFAULT_OPTIONS, ...userOptions }

  function _mix(...sources) {
    const result = {}
    for (const source of sources) {
      // Skip null or undefined sources
      if (source == null) continue
      merge(result, source, options)
    }
    return result
  }

  return _mix
}

/**
 * Recursively merges properties from input object into output object.
 * Handles property descriptors, prevents prototype pollution, and deep clones values.
 *
 * @param {Object} output - The target object to merge properties into.
 * @param {Object} input - The source object to merge properties from.
 * @param {Object} options - Merge options including array handling strategy.
 * @property {('replace'|'concat')} options.array - How to handle array merging.
 */
function merge(output, input, options) {
  const props = Object.keys(input)

  for (const prop of props) {
    // Prevents Prototype Pollution
    if (prop === '__proto__') continue

    const descriptor = Object.getOwnPropertyDescriptor(input, prop)
    const value = descriptor.value
    if (value) descriptor.value = cloneDescriptorValue(value)

    // If don't have prop => Define property
    if (!output[prop]) {
      Object.defineProperty(output, prop, descriptor)
      continue
    }

    // If have prop, but type is not object => Overwrite by redefining property
    if (typeof output[prop] !== 'object') {
      Object.defineProperty(output, prop, descriptor)
      continue
    }

    // Incoming non-object (primitive/null) => always overwrite
    if (descriptor.value == null || typeof descriptor.value !== 'object') {
      Object.defineProperty(output, prop, descriptor)
      continue
    }

    // If have prop, and incoming is Array => concat or replace
    if (objectType(descriptor.value) === '[object Array]') {
      output[prop] =
        options.array === 'replace'
          ? descriptor.value.slice()
          : output[prop].concat(descriptor.value)
      continue
    }

    // If have prop, but type is Object => Merge.
    merge(output[prop], descriptor.value, options)

    // Not sure if I have to deal with overwriting maps / dates. When it comes, I guess!
  }
}

/**
 * Creates a deep clone of a property descriptor value.
 * Handles arrays, objects, dates, maps, sets, and other types appropriately.
 *
 * @param {*} value - The value to clone.
 * @return {*} A deep clone of the input value.
 * @property {Array} value - Cloned recursively with all elements deep cloned.
 * @property {Object} value - Cloned with all properties and descriptors preserved.
 * @property {Date} value - Cloned as new Date with same time value.
 * @property {Map} value - Cloned with all entries deep cloned.
 * @property {Set} value - Cloned with all entries deep cloned.
 */
function cloneDescriptorValue(value) {
  // Arrays
  if (objectType(value) === '[object Array]') {
    const array = []
    for (let v of value) {
      v = cloneDescriptorValue(v)
      array.push(v)
    }
    return array
  }

  // Objects
  if (objectType(value) === '[object Object]') {
    const obj = {}
    const props = Object.keys(value)
    for (const prop of props) {
      const descriptor = Object.getOwnPropertyDescriptor(value, prop)
      if (descriptor.value)
        descriptor.value = cloneDescriptorValue(descriptor.value)
      Object.defineProperty(obj, prop, descriptor)
    }
    return obj
  }

  // Other Types of Objects
  if (objectType(value) === '[object Date]') {
    return new Date(value.getTime())
  }

  if (objectType(value) === '[object Map]') {
    const map = new Map()
    for (const entry of value) {
      map.set(entry[0], cloneDescriptorValue(entry[1]))
    }
    return map
  }

  if (objectType(value) === '[object Set]') {
    const set = new Set()
    for (const entry of value.entries()) {
      set.add(cloneDescriptorValue(entry[0]))
    }
    return set
  }

  // Types we don't need to clone or cannot clone.
  // Examples:
  // - Primitives don't need to clone
  // - Functions cannot clone
  return value
}

/**
 * Returns the precise object type of a value using Object.prototype.toString.
 *
 * @param {*} value - The value to get the type of.
 * @return {string} The object type string (e.g., '[object Array]', '[object Object]', '[object Date]').
 *
 * @example
 * objectType([]) // '[object Array]'
 * objectType({}) // '[object Object]'
 * objectType(new Date()) // '[object Date]'
 * objectType(null) // '[object Null]'
 */
function objectType(value) {
  return Object.prototype.toString.call(value)
}
