// zl-fetch/dist/index.js
function createRequestOptions(options = {}) {
  const opts = Object.assign({}, options);

  // Note: headers get mutated after setHeaders is called.
  // This is why we get the content type here to know what the user originally set.

  const headers = new Headers(options.headers);
  const userContentType = headers.get('content-type');

  opts.url = setUrl(opts);
  opts.method = setMethod(opts);
  opts.headers = setHeaders(opts);

  // If Content Type is set explicitly, we expect the user to pass in the appropriate data. So we don't treat the body
  opts.body = !userContentType ? setBody(opts) : opts.body;
  return opts
}

/**
 * Appends queries to URL
 * @param {Object} opts
 */
function setUrl(options) {
  const { url, queries, query, params, param } = options;

  // Merge queries, query, and params — treat them as the same item
  // So users don't have to remember singular or plural forms
  const q = Object.assign({}, queries, query, params, param);

  if (isEmptyObject(q)) return options.url

  const searchParams = new URLSearchParams(q);
  return `${url}?${searchParams.toString()}`
}

function isEmptyObject(obj) {
  return (
    obj && // 👈 null and undefined check
    Object.keys(obj).length === 0 &&
    Object.getPrototypeOf(obj) === Object.prototype
  )
}

function setMethod(options) {
  // Method set to GET by default unless otherwise specified
  const method = (options.method || 'get').toUpperCase();
  return method
}

// ========================
// Set Headers
// We set the headers depending on the request body and method
// ========================
function setHeaders(options) {
  let headers = new Headers(options.headers);
  headers = contentTypeHeader(options, headers);
  headers = authHeader(options, headers);

  return headers
}

function contentTypeHeader(options, headers) {
  // For preflight requests, we don't want to set headers.
  // This allows requests to remain simple.
  if (options.method === 'options') return headers

  // For GET requests, we also don't want to set headers.
  // This allows requests to remain simple.
  if (options.method === 'get') return headers

  // If a content type is aleady set, we return the content type as is.
  // This allows users to set their own content type.
  if (headers.get('content-type')) return headers

  // If the body is a string, we assume it's a query string.
  // So we set headers to Content-Type: application/x-www-form-urlencoded.
  if (typeof options.body === 'string') {
    headers.set('content-type', 'application/x-www-form-urlencoded');
    return headers
  }

  // If the body is an object and not FormData, we set it to `application/json`. Checking for FormData is important here because Form Data requires another content type.
  if (typeof options.body === 'object' && !isFormData(options.body)) {
    headers.set('content-type', 'application/json');
    return headers
  }

  // What's left here is FormData.
  // We don't set the content type here because fetch will set it automatically.
  return headers
}

// Sets the auth headers as necessary
function authHeader(options, headers) {
  // If no auth options, means we don't have to set Authorization headers
  const { auth } = options;
  if (!auth) return headers

  // Set Bearer authentication when user passes in a string into auth
  if (typeof auth === 'string') {
    headers.set('Authorization', `Bearer ${auth}`);
    return headers
  }

  // Set Basic Authentication headers when user passes in username and password into auth.
  const btoa = getBtoa();

  // Password field can be empty for implicit grant
  const { username, password = '' } = auth;

  if (!username) {
    throw new Error(
      'Username required to create Authorization Header for a Basic Authentication'
    )
  }

  const encodedValue = btoa(`${username}:${password}`);
  headers.set('Authorization', `Basic ${encodedValue}`);
  return headers
}

// Gets Btoa for creating basic auth.
function getBtoa() {
  if (typeof window !== 'undefined' && window.btoa) {
    return window.btoa
  }

  return function (string) {
    return Buffer.from(string).toString('base64')
  }
}

// ========================
// Set Body
// ========================
function setBody(options) {
  // Return empty body for preflight requests
  const method = options.method;
  if (['get', 'head', 'options'].includes(method)) return

  // If the body is form data, we return it as it is because users will have to set it up appropriately themselves.
  if (isFormData(options.body)) return options.body

  // If the body is a string, we assume it's a query string.
  // We ask users to use toQueryString to convert an object to a query string as they send in the request.
  if (typeof options.body === 'string') return options.body

  // If it's an object, we convert it to JSON
  if (typeof options.body === 'object') {
    return JSON.stringify(options.body)
  }

  // If the above conditions don't trigger, we return the body as is.
  // One example that will reach here is FormData
  return options.body
}

// Checks if body is FormData
// Only works on browsers.
// Returns false in Node because Node doesn't have FormData.
function isFormData(body) {
  if (typeof window === 'undefined') return false
  return body instanceof FormData
}

function parseJSON(value, defaultValue = {}) {
  try {
    return JSON.parse(value)
  } catch (e) {
    return defaultValue
  }
}

// Written by Claude AI and tested.
function omitEmpty(obj, shallow = false) {
  if (typeof obj !== 'object' || obj === null) return obj
  if (obj instanceof Date) return obj

  const result = Array.isArray(obj) ? [] : {};

  for (const [key, value] of Object.entries(obj)) {
    if (
      value === null ||
      value === undefined ||
      value === '' ||
      (typeof value === 'object' &&
        !(value instanceof Date) &&
        Object.keys(value).length === 0)
    ) {
      continue
    }
    result[key] = shallow ? value : omitEmpty(value);
  }

  return result
}

function getDefaultExportFromCjs (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

var require$$0 = {
	"100": "Continue",
	"101": "Switching Protocols",
	"102": "Processing",
	"103": "Early Hints",
	"200": "OK",
	"201": "Created",
	"202": "Accepted",
	"203": "Non-Authoritative Information",
	"204": "No Content",
	"205": "Reset Content",
	"206": "Partial Content",
	"207": "Multi-Status",
	"208": "Already Reported",
	"226": "IM Used",
	"300": "Multiple Choices",
	"301": "Moved Permanently",
	"302": "Found",
	"303": "See Other",
	"304": "Not Modified",
	"305": "Use Proxy",
	"307": "Temporary Redirect",
	"308": "Permanent Redirect",
	"400": "Bad Request",
	"401": "Unauthorized",
	"402": "Payment Required",
	"403": "Forbidden",
	"404": "Not Found",
	"405": "Method Not Allowed",
	"406": "Not Acceptable",
	"407": "Proxy Authentication Required",
	"408": "Request Timeout",
	"409": "Conflict",
	"410": "Gone",
	"411": "Length Required",
	"412": "Precondition Failed",
	"413": "Payload Too Large",
	"414": "URI Too Long",
	"415": "Unsupported Media Type",
	"416": "Range Not Satisfiable",
	"417": "Expectation Failed",
	"418": "I'm a Teapot",
	"421": "Misdirected Request",
	"422": "Unprocessable Entity",
	"423": "Locked",
	"424": "Failed Dependency",
	"425": "Too Early",
	"426": "Upgrade Required",
	"428": "Precondition Required",
	"429": "Too Many Requests",
	"431": "Request Header Fields Too Large",
	"451": "Unavailable For Legal Reasons",
	"500": "Internal Server Error",
	"501": "Not Implemented",
	"502": "Bad Gateway",
	"503": "Service Unavailable",
	"504": "Gateway Timeout",
	"505": "HTTP Version Not Supported",
	"506": "Variant Also Negotiates",
	"507": "Insufficient Storage",
	"508": "Loop Detected",
	"509": "Bandwidth Limit Exceeded",
	"510": "Not Extended",
	"511": "Network Authentication Required"
};

/*!
 * statuses
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2016 Douglas Christopher Wilson
 * MIT Licensed
 */

/**
 * Module dependencies.
 * @private
 */

var codes = require$$0;

/**
 * Module exports.
 * @public
 */

var statuses = status;

// status code to message map
status.message = codes;

// status message (lower-case) to code map
status.code = createMessageToStatusCodeMap(codes);

// array of status codes
status.codes = createStatusCodeList(codes);

// status codes for redirects
status.redirect = {
  300: true,
  301: true,
  302: true,
  303: true,
  305: true,
  307: true,
  308: true
};

// status codes for empty bodies
status.empty = {
  204: true,
  205: true,
  304: true
};

// status codes for when you should retry the request
status.retry = {
  502: true,
  503: true,
  504: true
};

/**
 * Create a map of message to status code.
 * @private
 */

function createMessageToStatusCodeMap (codes) {
  var map = {};

  Object.keys(codes).forEach(function forEachCode (code) {
    var message = codes[code];
    var status = Number(code);

    // populate map
    map[message.toLowerCase()] = status;
  });

  return map
}

/**
 * Create a list of all status codes.
 * @private
 */

function createStatusCodeList (codes) {
  return Object.keys(codes).map(function mapCode (code) {
    return Number(code)
  })
}

/**
 * Get the status code for given message.
 * @private
 */

function getStatusCode (message) {
  var msg = message.toLowerCase();

  if (!Object.prototype.hasOwnProperty.call(status.code, msg)) {
    throw new Error('invalid status message: "' + message + '"')
  }

  return status.code[msg]
}

/**
 * Get the status message for given code.
 * @private
 */

function getStatusMessage (code) {
  if (!Object.prototype.hasOwnProperty.call(status.message, code)) {
    throw new Error('invalid status code: ' + code)
  }

  return status.message[code]
}

/**
 * Get the status code.
 *
 * Given a number, this will throw if it is not a known status
 * code, otherwise the code will be returned. Given a string,
 * the string will be parsed for a number and return the code
 * if valid, otherwise will lookup the code assuming this is
 * the status message.
 *
 * @param {string|number} code
 * @returns {number}
 * @public
 */

function status (code) {
  if (typeof code === 'number') {
    return getStatusMessage(code)
  }

  if (typeof code !== 'string') {
    throw new TypeError('code must be a number or string')
  }

  // '403'
  var n = parseInt(code, 10);
  if (!isNaN(n)) {
    return getStatusMessage(n)
  }

  return getStatusCode(code)
}

var statuses$1 = /*@__PURE__*/getDefaultExportFromCjs(statuses);

// Server Sent Events (SSE) 

/**
 * Parses a single SSE message into a structured object
 * @param {string} message - The raw SSE message
 * @returns {Object|null} The parsed message or null if invalid
 * @property {string} [event] - The event type if specified
 * @property {string|Object} data - The message data (parsed as JSON if possible)
 * @property {string} [id] - The message ID if specified
 * @property {number} [retry] - The retry interval if specified
 */
function parseSSE(message) {
  // console.log('parsing SSE', message);
  const lines = message.split('\n');
  const result = {
    event: 'message', // Default event type
    data: '',
    id: null,
    retry: null
  };

  for (const line of lines) {
    if (line.startsWith('data:')) {
      const data = line.slice(5).trim();
      // Try to parse as JSON
      try { result.data = JSON.parse(data); } 
      // If not JSON, use as string
      catch { result.data = data; }
      continue
    } 
    
    if (line.startsWith('event:')) {
      result.event = line.slice(6).trim();
      continue
    } 
    
    if (line.startsWith('id:')) {
      result.id = line.slice(3).trim();
      continue
    } 
    
    if (line.startsWith('retry:')) {
      result.retry = parseInt(line.slice(6).trim(), 10);
      continue
    }
  }


  const ret = omitEmpty(result);
  // Only return if we have data
  return ret.data ? ret : null
}

/**
 * Handles the response from a fetch request and parses it according to content type
 * @param {Response} response - The fetch Response object
 * @param {Object} [options] - Configuration options
 * @param {Function} [options.customResponseParser] - Custom function to parse the response
 * @returns {Promise<Object>|Response} Returns either a parsed response object or the raw response if customResponseParser is used
 */
function handleResponse(response, options) {
  // Lets user use custom response parser because some people want to do so.
  // See https://github.com/zellwk/zl-fetch/issues/2
  if (options?.customResponseParser) return response
  const type = getResponseType(response, options);
  return parseResponse(response, { ...options, type })
}

/**
 * Formats fetch errors into a standardized zlFetch error format
 * @param {Error} error - The error object from the fetch request
 * @returns {Promise<Object>} Returns a rejected promise with the formatted error
 */
function handleError(error) {
  if (error.message === 'Failed to fetch') {
    /* eslint-disable */
    return Promise.reject({ error })
    /* eslint-enable */
  }
  return Promise.reject(error)
}

// ========================
// Internal Functions
// ========================
/**
 * Determines the type of response based on content-type and other headers
 * @param {Response} response - The fetch Response object
 * @returns {string|null} Returns the response type ('json', 'text', 'blob', 'formData', 'sse', 'chunked', 'stream') or null for 204 No Content
 * @throws {Error} Throws an error if the content-type is not supported
 */
function getResponseType(response, options) {
  const contentType = response.headers.get('content-type');
  const contentLength = response.headers.get('content-length');
  const transferEncoding = response.headers.get('Transfer-Encoding');

  if (!contentType) return null // Handles 204 No Content

  if (options.stream === true) {
    // Streaming response types
    if (contentType === 'text/event-stream') return 'sse'
    if (transferEncoding === 'chunked') return 'chunked'
    if (!contentLength) return 'stream'
  } else {
    if (contentType.includes('json')) return 'json'
    if (contentType.includes('text')) return 'text'
    if (contentType.includes('blob')) return 'blob'
    if (contentType.includes('x-www-form-urlencoded')) return 'formData'
  }

  // Need to check for FormData, Blob and ArrayBuffer content types
  throw new Error(`zlFetch does not support content-type ${contentType} yet`)
}

/**
 * Parses the response body according to its content type
 * @param {Response} response - The fetch Response object
 * @param {Object} options - Configuration options
 * @param {string} options.type - The type of response to parse
 * @returns {Promise<Object>} Returns a promise that resolves to the parsed response
 */
async function parseResponse(response, options) {
  // Parse formData into JavaScript object
  if (options.type === 'formData') {
    let body = await response.text();
    const query = new URLSearchParams(body);
    body = Object.fromEntries(query);
    return createOutput({ response, body, options })
  }

  if (options.type === 'stream') {
    const body = response.body;
    return createOutput({ response, body, options })
  }

  if (options.type === 'sse') {
    const body = handleSSEStream(response);
    return createOutput({ response, body, options })
  }

  if (options.type === 'chunked') {
    const body = handleChunkedStream(response);
    return createOutput({ response, body, options })
  }

  const body = await response[options.type]();
  return createOutput({ response, body, options })
}

/**
 * Creates a standardized output object for fetch responses
 * @param {Object} params - The parameters object
 * @param {Response} params.response - The fetch Response object
 * @param {*} params.body - The parsed response body
 * @param {Object} params.options - Configuration options
 * @param {boolean} [params.options.returnError=false] - If true, returns both success and error in a single object
 * @returns {Promise<Object>} Returns a promise that resolves to a standardized response object
 */
function createOutput({ response, body, options }) {
  const headers = getHeaders(response);
  const returnValue = {
    body,
    headers,
    response,
    status: response.status,
    statusText: response.statusText,
    abort: options.abortController.abort.bind(options.abortController),
  };

  // Resolves if successful response
  // Rejects if unsuccessful response
  if (!options.returnError) {
    return response.ok
      ? Promise.resolve(returnValue)
      : Promise.reject(returnValue)
  }

  // Returns both successful and unsuccessful response
  if (options.returnError) {
    let data;
    let error;

    if (response.ok) {
      data = returnValue;
      error = null;
    } else {
      data = null;
      error = returnValue;
    }

    return Promise.resolve({
      response: data,
      error,
    })
  }
}

/**
 * Extracts headers from the response object
 * @param {Response} response - The fetch Response object
 * @returns {Object} Returns an object containing all response headers
 */
function getHeaders(response) {
  return response.headers.entries
    ? getBrowserFetchHeaders(response)
    : getNodeFetchHeaders(response)
}

/**
 * Extracts headers from a browser fetch response
 * @param {Response} response - The fetch Response object
 * @returns {Object} Returns an object containing all response headers
 */
function getBrowserFetchHeaders(response) {
  const headers = {};
  for (const [header, value] of response.headers.entries()) {
    headers[header] = value;
  }
  return headers
}

/**
 * Extracts headers from a Node.js fetch response
 * @param {Response} response - The fetch Response object
 * @returns {Object} Returns an object containing all response headers
 */
function getNodeFetchHeaders(response) {
  const headers = {};
  const h = response.headers._headers;
  for (const header in h) {
    headers[header] = h[header].join('');
  }
  return headers
}

/**
 * Handles Server-Sent Events (SSE) stream responses
 * @param {Response} response - The fetch Response object
 * @returns {ReadableStream} Returns a ReadableStream that emits parsed SSE messages
 */
function handleSSEStream(response) {
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  return new ReadableStream({
    async start(controller) {
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break

          // Decode the chunk and add to buffer
          buffer += decoder.decode(value, { stream: true });
          // Process complete messages
          const messages = buffer.split('\n\n');
          buffer = messages.pop(); // Keep the last incomplete message in buffer

          for (const message of messages) {
            if (message.trim() === '') continue

            const parsedMessage = parseSSE(message);
            if (parsedMessage) controller.enqueue(parsedMessage);
          }
        }
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  })
}

/**
 * Handles chunked stream responses
 * @param {Response} response - The fetch Response object
 * @returns {ReadableStream} Returns a ReadableStream that emits parsed chunks
 */
function handleChunkedStream(response) {
  const reader = response.body.getReader();
  const decoder = new TextDecoder();

  return new ReadableStream({
    async start(controller) {
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break

          // Decode the chunk
          let chunk =
            typeof value === 'string'
              ? value
              : decoder.decode(value, { stream: true });

          chunk = chunk.trim();
          try {
            chunk = JSON.parse(chunk);
          } catch (error) {
            // Do nothing
          }
          controller.enqueue(chunk);
        }
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  })
}

/**
 * Main Fetch Function
 * @param {string} url - The endpoint URL to fetch from
 * @param {object} [options] - zlFetch options
 * @param {string} [options.method='GET'] - HTTP method (GET, POST, PUT, PATCH, DELETE)
 * @param {object} [options.query] - Query parameters object (alternative to queries)
 * @param {object} [options.queries] - Query parameters object (alternative to query)
 * @param {object} [options.params] - Query parameters object (alternative to query/queries)
 * @param {object} [options.param] - Query parameters object (alternative to query/queries/params)
 * @param {object} [options.headers] - HTTP headers to send with the request
 * @param {object|string|FormData} [options.body] - Request body. Can be an object (JSON), string (form-urlencoded), or FormData
 * @param {string|object} [options.auth] - Authentication information. String for Bearer token, object for Basic auth
 * @param {boolean} [options.debug=false] - When true, includes debug information in the response
 * @param {boolean} [options.returnError=false] - When true, returns error object instead of rejecting
 * @param {boolean} [options.customResponseParser=false] - When true, returns raw response without parsing
 * @returns {Promise<object>} A promise that resolves to the response object containing:
 *   - {object} headers - Response headers
 *   - {*} body - Parsed response body (JSON, text, or blob)
 *   - {number} status - HTTP status code
 *   - {string} statusText - HTTP status text
 *   - {Response} response - Original fetch Response object
 *   - {object} [debug] - Debug information (only if debug option is true)
 *   - {object} [error] - Error object (only if returnError is true)
 *   - {function} abort - Aborts the request
 * @throws {Error} When the request fails and returnError is false
 */

function coreFetch(url, options = {}) {
  const abortController = options.controller || new AbortController();
  const signal = options.signal || abortController.signal;
  const instance = fetchInstance({ url, ...options, abortController, signal });
  instance.abort = () => abortController.abort();

  return instance
}
// ========================
// Internal Functions
// ========================
async function fetchInstance(options) {
  const requestOptions = createRequestOptions(options);

  // Remove options that are not native to a fetch request
  delete requestOptions.fetch;
  delete requestOptions.queries;
  delete requestOptions.query;
  delete requestOptions.params;
  delete requestOptions.param;
  delete requestOptions.auth;
  delete requestOptions.debug;
  delete requestOptions.returnError;

  // Performs the fetch request
  return fetch(requestOptions.url, requestOptions)
    .then(response => handleResponse(response, options))
    .then(response => {
      if (options.signal) return response
      return response
    })
    .then(response => {
      if (!options.debug) return response
      return { ...response, debug: debugHeaders(requestOptions) }
    })
    .catch(handleError)
}

function debugHeaders(requestOptions) {
  const clone = Object.assign({}, requestOptions);
  const headers = {};
  for (const [header, value] of clone.headers) {
    headers[header] = value;
  }
  clone.headers = headers;
  return clone
}

// so it will be captured by ts

/**
 * @param {string} url - endpoint
 * @param {object} [options] - zlFetch options
 * @param {object} [options.query] - query Object
 * @param {object} [options.queries] - query Object
 * @param {object} [options.params] - query Object
 * @param {object} [options.param] - query Object
 * @param {object} [options.headers] - HTTP headers
 * @param {string} [options.auth] - Authentication information
 * @param {string} [options.debug] - Logs the request options for debugging
 * @param {string} [options.returnError] - Returns the error instead of rejecting it
 * @param {string} [options.customResponseParser] - Use a custome response parser
 */
coreFetch.get = function (url, options) {
  return coreFetch(url, {
    ...options,
    method: 'get'
  })
};

/**
 * @param {string} url - endpoint
 * @param {object} [options] - zlFetch options
 * @param {object} [options.query] - query Object
 * @param {object} [options.queries] - query Object
 * @param {object} [options.params] - query Object
 * @param {object} [options.param] - query Object
 * @param {object} [options.headers] - HTTP headers
 * @param {object} [options.body] - Body content
 * @param {string} [options.auth] - Authentication information
 * @param {string} [options.debug] - Logs the request options for debugging
 * @param {string} [options.returnError] - Returns the error instead of rejecting it
 * @param {string} [options.customResponseParser] - Use a custome response parser
 */
coreFetch.post = function (url, options) {
  return coreFetch(url, {
    ...options,
    method: 'post'
  })
};

/**
 * @param {string} url - endpoint
 * @param {object} [options] - zlFetch options
 * @param {object} [options.query] - query Object
 * @param {object} [options.queries] - query Object
 * @param {object} [options.params] - query Object
 * @param {object} [options.param] - query Object
 * @param {object} [options.headers] - HTTP headers
 * @param {object} [options.body] - Body content
 * @param {string} [options.auth] - Authentication information
 * @param {string} [options.debug] - Logs the request options for debugging
 * @param {string} [options.returnError] - Returns the error instead of rejecting it
 * @param {string} [options.customResponseParser] - Use a custome response parser
 */
coreFetch.put = function (url, options) {
  return coreFetch(url, {
    ...options,
    method: 'put'
  })
};

/**
 * @param {string} url - endpoint
 * @param {object} [options] - zlFetch options
 * @param {object} [options.query] - query Object
 * @param {object} [options.queries] - query Object
 * @param {object} [options.params] - query Object
 * @param {object} [options.param] - query Object
 * @param {object} [options.headers] - HTTP headers
 * @param {object} [options.body] - Body content
 * @param {string} [options.auth] - Authentication information
 * @param {string} [options.debug] - Logs the request options for debugging
 * @param {string} [options.returnError] - Returns the error instead of rejecting it
 * @param {string} [options.customResponseParser] - Use a custome response parser
 */
coreFetch.patch = function (url, options) {
  return coreFetch(url, {
    ...options,
    method: 'patch'
  })
};

/**
 * @param {string} url - endpoint
 * @param {object} [options] - zlFetch options
 * @param {object} [options.query] - query Object
 * @param {object} [options.queries] - query Object
 * @param {object} [options.params] - query Object
 * @param {object} [options.param] - query Object
 * @param {object} [options.headers] - HTTP headers
 * @param {string} [options.auth] - Authentication information
 * @param {string} [options.debug] - Logs the request options for debugging
 * @param {string} [options.returnError] - Returns the error instead of rejecting it
 * @param {string} [options.customResponseParser] - Use a custome response parser
 */
coreFetch.delete = function (url, options) {
  return coreFetch(url, {
    ...options,
    method: 'delete'
  })
};

// Creates an instance of zlFetch to be used later
function create(baseURL, options) {
  const fn = function (...args) {
    const { url, newOptions } = normalize(args);
    return coreFetch(url, { ...options, ...newOptions })
  };

  // Create Shorthand methods
  const methods = ['get', 'post', 'put', 'patch', 'delete'];

  for (const method of methods) {
    fn[method] = function (...args) {
      const { url, newOptions } = normalize(args);
      return coreFetch(url, { ...options, ...newOptions, method })
    };
  }

  // Normalize the URL and options
  // Allows user to use the created zlFetch item without passing in further URLs.
  // Naming can be improved, but can't think of a better name for now
  function normalize(args = []) {
    const [arg1, arg2] = args;

    // This means no options are given. So we simply use the baseURL as the URL
    if (!arg1) return { url: baseURL }

    // If the firs argument is an object, it means there are options but the user didn't pass in a URL.
    if (typeof arg1 === 'object') return { url: baseURL, newOptions: arg1 }

    // The leftover possibility is that the first argument is a string.
    // In this case we need to make a new URL with this argument.
    const url = makeURL(baseURL, arg1);

    // Wwe need to check whether the second argument is an object or not. If arg2 is undefined, then we simply return the URL since there are no options
    if (!arg2) return { url }

    // The only possibility left is that arg2 is an object, which means there are new options.
    return { url, newOptions: arg2 }
  }

  return fn
}

// Joins the baseURL and endpoint.
// Uses a simple string concatenation instead of path.join so it works in the browser.
function makeURL(baseURL, url) {
  if (baseURL.endsWith('/') && url.startsWith('/')) {
    url = url.slice(1);
  }

  if (!baseURL.endsWith('/') && !url.startsWith('/')) {
    url = '/' + url;
  }

  return baseURL + url
}

/**
 * Creates an EventSource connection with automatic JSON parsing and event handling.
 * Works in both browser and Node.js environments with consistent behavior.
 *
 * EventSource automatically handles reconnection:
 * - Default retry interval is 3 seconds
 * - Server can override retry interval by sending `retry: <milliseconds>`
 * - Retry interval persists until changed by server or connection is closed
 *
 * @param {string} url - The URL to connect to for the EventSource
 * @param {Object} options - Configuration options
 * @param {boolean} [options.useFetch=false] - Force using fetch-based implementation even in browser
 * @param {number} [options.retry=3000] - Default retry interval in milliseconds (Node.js only)
 * @param {Function} [options.message] - Callback for 'message' events. Receives parsed JSON data or raw text if parsing fails
 * @param {Function} [options.open] - Callback for 'open' events. Receives parsed JSON data or raw text if parsing fails
 * @param {Function} [options.error] - Callback for 'error' events. Receives parsed JSON data or raw text if parsing fails
 * @param {Function} [options.close] - Callback for 'close' events. Called when the connection is closed, either by the server sending a 'close' event or manually. Receives parsed JSON data from the close event if available
 * @param {Function} [options.*] - Callbacks for any custom events. Each callback receives parsed JSON data or raw text if parsing fails
 * @param {Object} [fetchOptions] - Options to pass to fetch/zlFetch when using fetch-based implementation
 * @returns {EventSource|Promise<ReadableStream>} In browser: EventSource instance that can be used to close the connection. In Node.js: Promise that resolves to a ReadableStream
 *
 * @example
 * // Basic usage with standard events
 * const source = zlEventSource('http://localhost:3000/stream', {
 *   message: (data) => console.log('Message:', data),
 *   open: (data) => console.log('Connected:', data),
 *   error: (error) => console.log('Error:', error),
 *   close: (data) => console.log('Connection closed:', data)
 * })
 *
 * // Close connection when done (browser only)
 * source.close()
 *
 * @example
 * // With custom events
 * zlEventSource('http://localhost:3000/stream', {
 *   status: (data) => console.log('Status:', data),
 *   update: (data) => console.log('Update:', data)
 * })
 *
 * @example
 * // Using fetch implementation with custom options
 * const stream = await zlEventSource(
 *   'http://localhost:3000/stream',
 *   {
 *     useFetch: true,
 *     retry: 5000, // 5 second retry interval
 *     message: (data) => console.log('Message:', data)
 *   },
 *   {
 *     headers: { 'Authorization': 'Bearer token' },
 *     credentials: 'include'
 *   }
 * )
 */
function zlEventSource(
  url,
  { useFetch = false, ...options } = {},
  fetchOptions = {},
) {
  if (useFetch || typeof EventSource === 'undefined')
    return zlNodeEventSource(url, options, fetchOptions)
  return browserEventSource(url, options)
}

/**
 * Creates an EventSource connection in the browser environment.
 * Uses the native EventSource API with automatic JSON parsing.
 *
 * @private
 * @param {string} url - The URL to connect to for the EventSource
 * @param {Object} options - Configuration options containing event callbacks
 * @param {Object} options.callbacks - Object containing event name to callback function mappings
 * @returns {EventSource} The EventSource instance that can be used to close the connection
 */
function browserEventSource(url, { close, ...callbacks } = {}) {
  const evtSrc = new EventSource(url);
  const listeners = new Map();

  // Set up event handlers using addEventListener
  for (const [event, callback] of Object.entries(callbacks)) {
    const handler = event => {
      const data = parseJSON(event.data);
      callback(data);
    };
    evtSrc.addEventListener(event, handler);
    listeners.set(event, handler);
  }

  // Handle close event
  evtSrc.addEventListener('close', event => {
    const data = parseJSON(event.data);

    // Remove all registered listeners
    for (const [event, handler] of listeners) {
      evtSrc.removeEventListener(event, handler);
    }
    listeners.clear();

    // Actually close the connection
    evtSrc.close();

    // Call the close callback if provided
    if (close) close(data);
  });

  return evtSrc
}

/**
 * Creates an EventSource connection in the Node.js environment.
 * Uses zlFetch to handle the connection and stream processing.
 *
 * @private
 * @param {string} url - The URL to connect to for the EventSource
 * @param {Object} options - Configuration options
 * @param {number} [options.retry=3000] - Default retry interval in milliseconds
 * @param {Object} options.callbacks - Object containing event name to callback function mappings
 * @param {Object} [fetchOptions] - Options to pass to zlFetch for the connection
 * @returns {Promise<ReadableStream>} Promise that resolves to a ReadableStream for the SSE connection
 */
function zlNodeEventSource(
  url,
  { retry = 3000, ...callbacks } = {},
  fetchOptions = {},
) {
  let shouldStop = false;
  let stream;

  const evtSource = connect();
  evtSource.close = () => {
    shouldStop = true;
    stream.abort();
  };

  async function connect() {
    stream = coreFetch(url, { ...fetchOptions, stream: true });

    return stream
      .then(async stream => {
        for await (const chunk of stream.body) {
          if (shouldStop) break
          const cb = callbacks[chunk.event];
          if (cb) cb(chunk.data);
          if (chunk.event === 'close') evtSource.close();
        }
      })
      .catch(error => {
        const cb = callbacks.error;
        if (cb) cb(error);
        if (!shouldStop) {
          return setTimeout(connect, retry)
        }
      })
  }

  return evtSource
}

/**
 * Converts Form Data into a plain JavaScript object
 * @param {FormData} formData - The FormData object to convert
 * @returns {Object} A plain JavaScript object with form data key-value pairs
 * @example
 * const formData = new FormData()
 * formData.append('name', 'John')
 * formData.append('email', 'john@example.com')
 * const obj = toObject(formData)
 * // Returns: { name: 'John', email: 'john@example.com' }
 */
function toObject(formData) {
  const obj = {};
  for (const data of formData) {
    obj[data[0]] = data[1].trim();
  }
  return obj
}

/**
 * Converts a plain JavaScript object into a URL query string
 * @param {Object} object - The object to convert to query string
 * @returns {string} A URL-encoded query string
 * @example
 * const obj = { name: 'John', age: '30' }
 * const queryString = toQueryString(obj)
 * // Returns: 'name=John&age=30'
 */
function toQueryString(object) {
  const searchParams = new URLSearchParams(object);
  return searchParams.toString()
}

// When rejecting API errors, use this to make them easier to read.
// Also doubles up as a way to standardize error messages and status codes.

/**
 * Standardizes and formats error objects for API rejections
 * @param {Object} error - The error object to format
 * @param {number} error.status - HTTP status code
 * @param {string} [error.statusText] - HTTP status text
 * @param {*} [error.body] - Response body if available
 * @param {string} [error.message] - Error message if available
 * @returns {Promise<never>} A rejected promise with a standardized error object
 * @example
 * reject({ status: 404, statusText: 'Not Found', body: { message: 'Resource not found' } })
 * // Returns: Promise.reject({ status: 404, statusText: 'Not Found', body: { message: 'Resource not found' } })
 */
function reject(error) {
  const err = {
    status: error.status,
    statusText: error.statusText || statuses$1(error.status),
  };

  if (error.body) err.body = error.body;
  if (error.message) err.message = error.message;

  return Promise.reject(err)
}

/**
 * Reads and processes a response stream
 * @param {Response} response - The fetch Response object to read
 * @returns {Promise<any>} A promise that resolves with the processed stream data
 * @see {@link handleChunkedStream} for implementation details
 */
function readStream(response) {
  return handleChunkedStream(response)
}

const fn = coreFetch;
const createZlFetch = create;

export { create, createZlFetch, fn as default, readStream, reject, toObject, toQueryString, zlEventSource, zlNodeEventSource };
