// @splendidlabz/utils/dist/esm/lib/strings/markdown.js
import markdownit from "markdown-it";
import anchor from "markdown-it-anchor";
const md = markdownit({
  html: true,
  linkify: true,
  typographer: true
});
md.use(anchor, {
  permalink: false,
  slugify: (s) => s.toLowerCase().replace(/\s+/g, "-").replace(/[^\w-]/g, "")
});
function markdown(content, options = {}) {
  const defaultOptions = { inline: false };
  const opts = { ...defaultOptions, ...options };
  const { content: treated } = treatMarkdownWhitespace(content, opts);
  if (opts.inline) return md.renderInline(treated);
  return md.render(treated);
}
function treatMarkdownWhitespace(content, { inline } = {}) {
  if (!content) return { content: "", inline };
  const lines = content.split("\n");
  if (lines[0] !== "" && inline) return { content, inline: true };
  const firstLineWithLeadingSpacesIndex = lines.findIndex(
    (line) => line.match(/^\s+/)
  );
  const firstLineWithLeadingSpaces = lines[firstLineWithLeadingSpacesIndex];
  const indentation = firstLineWithLeadingSpaces?.match(/^\s+/)?.[0]?.length ?? 0;
  const retContent = lines.map((line, index) => {
    if (index < firstLineWithLeadingSpacesIndex) return line + "\n";
    return line.slice(indentation) + "\n";
  }).join("");
  return {
    content: retContent,
    inline
  };
}
export {
  markdown,
  treatMarkdownWhitespace
};
