// @splendidlabz/utils/dist/esm/dom/bounding-box.js
function boundingBox(element, { containsPadding = true } = {}) {
  const box = element.getBoundingClientRect();
  const ret = {};
  for (const prop in box) ret[prop] = box[prop];
  if (!containsPadding) {
    const style = getComputedStyle(element);
    const paddingTop = parseFloat(style.paddingTop);
    const paddingRight = parseFloat(style.paddingRight);
    const paddingBottom = parseFloat(style.paddingBottom);
    const paddingLeft = parseFloat(style.paddingLeft);
    ret.x = ret.x + paddingLeft;
    ret.y = ret.y + paddingTop;
    ret.width = ret.width - paddingLeft - paddingRight;
    ret.height = ret.height - paddingTop - paddingBottom;
    ret.top = ret.top + paddingTop;
    ret.bottom = ret.bottom - paddingBottom;
    ret.right = ret.right - paddingRight;
    ret.left = ret.left + paddingLeft;
  }
  ret.xCenter = (ret.left + ret.right) / 2;
  ret.yCenter = (ret.top + ret.bottom) / 2;
  return ret;
}
function boundingBoxRelativeToAncestor({ element, ancestorElement }) {
  ancestorElement = ancestorElement || element.parentElement;
  const box = boundingBox(element);
  const ancestorBox = boundingBox(ancestorElement, {
    containsPadding: false
  });
  return {
    top: box.top - ancestorBox.top,
    right: box.right - ancestorBox.right,
    bottom: box.bottom - ancestorBox.bottom,
    left: box.left - ancestorBox.left,
    width: box.width,
    height: box.height,
    xCenter: (box.left + box.right) / 2 - ancestorBox.left,
    yCenter: (box.top + box.bottom) / 2 - ancestorBox.top
  };
}
export {
  boundingBox,
  boundingBoxRelativeToAncestor
};
