// @splendidlabz/utils/dist/esm/dom/get-element.js
const astroNodes = ["ASTRO-SLOT", "ASTRO-ISLAND"];
function getNodeType(node) {
  if (node instanceof Element) return "element";
  if (node instanceof NodeList) return "nodelist";
  if (Array.isArray(node)) return "array";
}
function getElement(selector) {
  if (!selector) return;
  if (selector instanceof Element) return selector;
  return document.querySelector(selector);
}
function getChildrenElements(node) {
  let children = node.children;
  if (children[0]?.nodeName === "ASTRO-SLOT") children = children[0].children;
  return Array.from(children);
}
function getParentElement(element) {
  const parent = element.parentElement;
  if (astroNodes.includes(parent.nodeName)) return getParentElement(parent);
  return parent;
}
function getSiblingElements(element) {
  return Array.from(element.parentElement.children).filter(
    (child) => child !== element
  );
}
function getSelfIndex(element) {
  return Array.from(element.parentElement.children).indexOf(element);
}
function isAncestor(ancestor, element, { searchLevels = 0 } = {}) {
  if (ancestor === element) return true;
  let parent = element.parentElement;
  if (searchLevels === 0) {
    while (parent) {
      if (ancestor === parent) return true;
      parent = parent.parentElement;
    }
  } else {
    for (let i = 0; i < searchLevels; i++) {
      if (ancestor === parent) return true;
      parent = parent.parentElement;
    }
  }
  return false;
}
function getAncestorWithSiblings(element, { limit = false } = {}) {
  const ancestor = element.parentElement;
  const siblings = getSiblingElements(ancestor);
  if (siblings.length) return siblings;
  if (!ancestor) return null;
  if (limit === false) return getAncestorWithSiblings(ancestor);
  if (typeof limit === "number" && limit > 0) {
    return getAncestorWithSiblings(ancestor, { limit: limit - 1 });
  }
  if (limit === 0) return null;
}
export {
  getAncestorWithSiblings,
  getChildrenElements,
  getElement,
  getNodeType,
  getParentElement,
  getSelfIndex,
  getSiblingElements,
  isAncestor
};
