// @splendidlabz/utils/dist/cjs/lib/strings/markdown.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var markdown_exports = {};
__export(markdown_exports, {
  markdown: () => markdown,
  treatMarkdownWhitespace: () => treatMarkdownWhitespace
});
module.exports = __toCommonJS(markdown_exports);
var import_markdown_it = __toESM(require("markdown-it"), 1);
var import_markdown_it_anchor = __toESM(require("markdown-it-anchor"), 1);
const md = (0, import_markdown_it.default)({
  html: true,
  linkify: true,
  typographer: true
});
md.use(import_markdown_it_anchor.default, {
  permalink: false,
  slugify: (s) => s.toLowerCase().replace(/\s+/g, "-").replace(/[^\w-]/g, "")
});
function markdown(content, options = {}) {
  const defaultOptions = { inline: false };
  const opts = { ...defaultOptions, ...options };
  const { content: treated } = treatMarkdownWhitespace(content, opts);
  if (opts.inline) return md.renderInline(treated);
  return md.render(treated);
}
function treatMarkdownWhitespace(content, { inline } = {}) {
  if (!content) return { content: "", inline };
  const lines = content.split("\n");
  if (lines[0] !== "" && inline) return { content, inline: true };
  const firstLineWithLeadingSpacesIndex = lines.findIndex(
    (line) => line.match(/^\s+/)
  );
  const firstLineWithLeadingSpaces = lines[firstLineWithLeadingSpacesIndex];
  const indentation = firstLineWithLeadingSpaces?.match(/^\s+/)?.[0]?.length ?? 0;
  const retContent = lines.map((line, index) => {
    if (index < firstLineWithLeadingSpacesIndex) return line + "\n";
    return line.slice(indentation) + "\n";
  }).join("");
  return {
    content: retContent,
    inline
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  markdown,
  treatMarkdownWhitespace
});
