// @splendidlabz/utils/src/node/file-cache.js
import fs from 'fs/promises'
import { glob } from 'glob'
import path from 'node:path'
import { sort } from '../lib/arrays/index.js'

/**
 * Resolves a path to an absolute path, handling both absolute and relative paths
 * @param {string} dirPath - Absolute or relative path to directory
 * @returns {string} Absolute path
 */
function resolveDirPath(dirPath) {
  return path.isAbsolute(dirPath) ? dirPath : path.join(process.cwd(), dirPath)
}

/**
 * Creates a file cache instance for managing cached files in a directory
 * @param {string} dir - Absolute or relative path to cache directory
 * @returns {Promise<object>} Cache instance
 * @property {function(string): Promise<number>} cacheAt - Returns the cached file's modified timestamp, or directory's last modified time if no file specified
 * @property {function(string): string} getAbsPath - Returns absolute path for a file in cache
 * @property {function(string): Promise<boolean>} hasFile - Checks if a file exists in cache
 * @property {function(string): Promise<Buffer|undefined>} loadFile - Loads file contents from cache
 * @property {function(string, string|Buffer): Promise<void>} saveFile - Saves data to cache file
 * @property {function(string): Promise<object>} loadJSON - Loads and parses JSON from cache file
 * @property {function(string, object): Promise<void>} saveJSON - Saves object as JSON to cache file
 */
export async function createCache(dir) {
  const absDirpath = resolveDirPath(dir)
  await fs.mkdir(absDirpath, { recursive: true })

  return {
    async cacheAt(file) {
      if (file) {
        // Return the cached file's modified timestamp
        const absFilepath = path.join(absDirpath, file)
        try {
          const stat = await fs.stat(absFilepath)
          return stat.mtimeMs
        } catch (error) {
          return 0
        }
      } else {
        // Return the last modified timestamp of the cache directory
        return getLastModifiedTime(absDirpath + '/**/*')
      }
    },

    getAbsPath(file) {
      return path.join(absDirpath, file)
    },

    async hasFile(file) {
      const absFilepath = path.join(absDirpath, file)
      try {
        await fs.stat(absFilepath)
        return true
      } catch (e) {
        return false
      }
    },

    async loadFile(file) {
      try {
        const absFilepath = path.join(absDirpath, file)
        return fs.readFile(absFilepath)
      } catch (error) {
        console.warn(`File not found: ${file}`)
      }
    },

    async saveFile(file, data) {
      const absfilepath = path.join(absDirpath, file)
      await fs.writeFile(absfilepath, data)
    },

    async loadJSON(file) {
      const data = await this.loadFile(file)
      try {
        return JSON.parse(data.toString())
      } catch (err) {
        console.warn(`Error parsing JSON: ${file}`)
        return {}
      }
    },

    async saveJSON(file, data) {
      await this.saveFile(file, JSON.stringify(data))
    },
  }
}

/**
 * Legacy file cache API for managing a single cached file
 * @param {Object} options
 * @param {string} options.dir - Absolute or relative path to cache directory
 * @param {string} options.file - Filename to cache
 * @returns {Promise<{
 *   modifiedAt: number,
 *   load: () => Promise<Buffer>,
 *   save: (data: string|Buffer) => Promise<void>,
 *   loadJSON: () => Promise<object>,
 *   saveJSON: (data: object) => Promise<void>
 * }>}
 */
export async function fileCache({ dir, file }) {
  // Makes the path if it doesn't exist
  const absDirpath = resolveDirPath(dir)
  await fs.mkdir(absDirpath, { recursive: true })
  const cachePath = path.join(absDirpath, `${file}`)

  let modifiedAt = 0

  try {
    const stat = await fs.stat(cachePath)
    modifiedAt = stat.mtimeMs
  } catch (error) {} // Silent error

  return {
    modifiedAt,

    async load() {
      return fs.readFile(cachePath)
    },

    async save(data) {
      await fs.writeFile(cachePath, data)
    },

    async loadJSON() {
      const data = await fs.readFile(cachePath)
      return JSON.parse(data.toString())
    },
    async saveJSON(data) {
      await fs.writeFile(cachePath, JSON.stringify(data))
    },
  }
}

/**
 * Gets the most recent modification time of files matching the glob pattern
 * @param {string|string[]} globPath - Glob pattern(s) to match files
 * @returns {Promise<number>} Timestamp of most recently modified file, or 0 if no files found
 */
export async function getLastModifiedTime(globPath) {
  let files

  if (typeof globPath === 'string') {
    files = await glob(removeFirstSlash(globPath))
  }

  if (Array.isArray(globPath)) {
    const a = await Promise.all(
      globPath.map(path => glob(removeFirstSlash(path))),
    )
    files = a.flat()
  }

  const stats = await Promise.all(files.map(file => fs.stat(file)))
  const timestamps = stats.map(stat => Number(stat.mtimeMs))

  const sorted = sort(timestamps, { props: 'mtimeMs', reverse: true })
  return sorted[0] || 0
}

/**
 * @deprecated Use getLastModifiedTime instead
 * @param {string|string[]} globPath - Glob pattern(s) to match files
 * @returns {Promise<number>} Timestamp of most recently modified file, or 0 if no files found
 */
export async function getLatestModifiedTime(globPath) {
  console.warn(
    '`getLatestModifiedTime` has been renamed to `getLastModifiedTime` and will be deprecated in future versions. Please use `getLastModifiedTime` instead.',
  )
  return getLastModifiedTime(globPath)
}

/**
 * Removes leading slash from a path string
 * @param {string} x - Path string to process
 * @returns {string} Path without leading slash
 */
export function removeFirstSlash(x) {
  return x.startsWith('/') ? x.slice(1) : x
}
