// @splendidlabz/utils/dist/cjs/dom/observers/intersection-observer.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var intersection_observer_exports = {};
__export(intersection_observer_exports, {
  intersectionObserver: () => intersectionObserver
});
module.exports = __toCommonJS(intersection_observer_exports);
var import_events = require("../events.js");
var import_observer = require("./observer.js");
function intersectionObserver(target, options = {}) {
  const { callback, ...opts } = options;
  const observer = new IntersectionObserver(observerFn, opts);
  (0, import_observer.useObserverMethodOnTarget)(target, observer, "observe");
  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer });
      else (0, import_events.dispatchEvent)(target, "intersect", { entry, entries, observer });
    }
  }
  return {
    observer,
    observe(target2) {
      (0, import_observer.useObserverMethodOnTarget)(target2, observer, "observe", options);
    },
    unobserve(target2) {
      (0, import_observer.useObserverMethodOnTarget)(target2, observer, "unobserve");
    },
    takeRecords() {
      return observer.takeRecords();
    },
    disconnect() {
      const records = observer.takeRecords();
      observer.disconnect();
      if (records.length > 0) observerFn(records);
    },
    destroy() {
      observer.disconnect();
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  intersectionObserver
});
