// @splendidlabz/utils/dist/cjs/dom/observers/resize-observer.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var resize_observer_exports = {};
__export(resize_observer_exports, {
  resizeObserver: () => resizeObserver
});
module.exports = __toCommonJS(resize_observer_exports);
var import_events = require("../events.js");
var import_observer = require("./observer.js");
function resizeObserver(target, options) {
  const { callback, ...opts } = options;
  const observer = new ResizeObserver(observerFn);
  if (target === window || target === document)
    target = document.documentElement;
  (0, import_observer.useObserverMethodOnTarget)(target, observer, "observe", opts);
  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer });
      else
        (0, import_events.dispatchEvent)(entry.target, "resize-obs", { entry, entries, observer });
    }
  }
  return {
    observe(target2) {
      (0, import_observer.useObserverMethodOnTarget)(target2, observer, "observe", options);
    },
    unobserve(target2) {
      (0, import_observer.useObserverMethodOnTarget)(target2, observer, "unobserve");
    },
    disconnect: (_) => observer.disconnect(),
    destroy: (_) => observer.disconnect()
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  resizeObserver
});
