// @splendidlabz/utils/dist/cjs/node/file-cache.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var file_cache_exports = {};
__export(file_cache_exports, {
  createCache: () => createCache,
  fileCache: () => fileCache,
  getLastModifiedTime: () => getLastModifiedTime,
  getLatestModifiedTime: () => getLatestModifiedTime,
  removeFirstSlash: () => removeFirstSlash
});
module.exports = __toCommonJS(file_cache_exports);
var import_promises = __toESM(require("fs/promises"), 1);
var import_glob = require("glob");
var import_node_path = __toESM(require("node:path"), 1);
var import_arrays = require("../lib/arrays/index.js");
function resolveDirPath(dirPath) {
  return import_node_path.default.isAbsolute(dirPath) ? dirPath : import_node_path.default.join(process.cwd(), dirPath);
}
async function createCache(dir) {
  const absDirpath = resolveDirPath(dir);
  await import_promises.default.mkdir(absDirpath, { recursive: true });
  return {
    async cacheAt(file) {
      if (file) {
        const absFilepath = import_node_path.default.join(absDirpath, file);
        try {
          const stat = await import_promises.default.stat(absFilepath);
          return stat.mtimeMs;
        } catch (error) {
          return 0;
        }
      } else {
        return getLastModifiedTime(absDirpath + "/**/*");
      }
    },
    getAbsPath(file) {
      return import_node_path.default.join(absDirpath, file);
    },
    async hasFile(file) {
      const absFilepath = import_node_path.default.join(absDirpath, file);
      try {
        await import_promises.default.stat(absFilepath);
        return true;
      } catch (e) {
        return false;
      }
    },
    async loadFile(file) {
      try {
        const absFilepath = import_node_path.default.join(absDirpath, file);
        return import_promises.default.readFile(absFilepath);
      } catch (error) {
        console.warn(`File not found: ${file}`);
      }
    },
    async saveFile(file, data) {
      const absfilepath = import_node_path.default.join(absDirpath, file);
      await import_promises.default.writeFile(absfilepath, data);
    },
    async loadJSON(file) {
      const data = await this.loadFile(file);
      try {
        return JSON.parse(data.toString());
      } catch (err) {
        console.warn(`Error parsing JSON: ${file}`);
        return {};
      }
    },
    async saveJSON(file, data) {
      await this.saveFile(file, JSON.stringify(data));
    }
  };
}
async function fileCache({ dir, file }) {
  const absDirpath = resolveDirPath(dir);
  await import_promises.default.mkdir(absDirpath, { recursive: true });
  const cachePath = import_node_path.default.join(absDirpath, `${file}`);
  let modifiedAt = 0;
  try {
    const stat = await import_promises.default.stat(cachePath);
    modifiedAt = stat.mtimeMs;
  } catch (error) {
  }
  return {
    modifiedAt,
    async load() {
      return import_promises.default.readFile(cachePath);
    },
    async save(data) {
      await import_promises.default.writeFile(cachePath, data);
    },
    async loadJSON() {
      const data = await import_promises.default.readFile(cachePath);
      return JSON.parse(data.toString());
    },
    async saveJSON(data) {
      await import_promises.default.writeFile(cachePath, JSON.stringify(data));
    }
  };
}
async function getLastModifiedTime(globPath) {
  let files;
  if (typeof globPath === "string") {
    files = await (0, import_glob.glob)(removeFirstSlash(globPath));
  }
  if (Array.isArray(globPath)) {
    const a = await Promise.all(
      globPath.map((path2) => (0, import_glob.glob)(removeFirstSlash(path2)))
    );
    files = a.flat();
  }
  const stats = await Promise.all(files.map((file) => import_promises.default.stat(file)));
  const timestamps = stats.map((stat) => Number(stat.mtimeMs));
  const sorted = (0, import_arrays.sort)(timestamps, { props: "mtimeMs", reverse: true });
  return sorted[0] || 0;
}
async function getLatestModifiedTime(globPath) {
  console.warn(
    "`getLatestModifiedTime` has been renamed to `getLastModifiedTime` and will be deprecated in future versions. Please use `getLastModifiedTime` instead."
  );
  return getLastModifiedTime(globPath);
}
function removeFirstSlash(x) {
  return x.startsWith("/") ? x.slice(1) : x;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createCache,
  fileCache,
  getLastModifiedTime,
  getLatestModifiedTime,
  removeFirstSlash
});
