// @splendidlabz/utils/dist/cjs/node/pkce.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pkce_exports = {};
__export(pkce_exports, {
  PKCE: () => PKCE
});
module.exports = __toCommonJS(pkce_exports);
var import_node_crypto = __toESM(require("node:crypto"), 1);
var import_random_string = require("./random-string.js");
async function PKCE() {
  const codeVerifier = await (0, import_random_string.randomString)();
  const codeChallenge = await getCodeChallenge(codeVerifier);
  return {
    state: await (0, import_random_string.randomString)(),
    code_verifier: codeVerifier,
    code_challenge: codeChallenge,
    code_challenge_method: "S256"
  };
}
async function getCodeChallenge(verifier) {
  const hash = import_node_crypto.default.createHash("sha256").update(verifier).digest();
  return hash.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PKCE
});
