// @splendidlabz/utils/dist/cjs/dom/bounding-box.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bounding_box_exports = {};
__export(bounding_box_exports, {
  boundingBox: () => boundingBox,
  boundingBoxRelativeToAncestor: () => boundingBoxRelativeToAncestor
});
module.exports = __toCommonJS(bounding_box_exports);
function boundingBox(element, { containsPadding = true } = {}) {
  const box = element.getBoundingClientRect();
  const ret = {};
  for (const prop in box) ret[prop] = box[prop];
  if (!containsPadding) {
    const style = getComputedStyle(element);
    const paddingTop = parseFloat(style.paddingTop);
    const paddingRight = parseFloat(style.paddingRight);
    const paddingBottom = parseFloat(style.paddingBottom);
    const paddingLeft = parseFloat(style.paddingLeft);
    ret.x = ret.x + paddingLeft;
    ret.y = ret.y + paddingTop;
    ret.width = ret.width - paddingLeft - paddingRight;
    ret.height = ret.height - paddingTop - paddingBottom;
    ret.top = ret.top + paddingTop;
    ret.bottom = ret.bottom - paddingBottom;
    ret.right = ret.right - paddingRight;
    ret.left = ret.left + paddingLeft;
  }
  ret.xCenter = (ret.left + ret.right) / 2;
  ret.yCenter = (ret.top + ret.bottom) / 2;
  return ret;
}
function boundingBoxRelativeToAncestor({ element, ancestorElement }) {
  ancestorElement = ancestorElement || element.parentElement;
  const box = boundingBox(element);
  const ancestorBox = boundingBox(ancestorElement, {
    containsPadding: false
  });
  return {
    top: box.top - ancestorBox.top,
    right: box.right - ancestorBox.right,
    bottom: box.bottom - ancestorBox.bottom,
    left: box.left - ancestorBox.left,
    width: box.width,
    height: box.height,
    xCenter: (box.left + box.right) / 2 - ancestorBox.left,
    yCenter: (box.top + box.bottom) / 2 - ancestorBox.top
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  boundingBox,
  boundingBoxRelativeToAncestor
});
