// @splendidlabz/utils/dist/cjs/dom/focusable.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var focusable_exports = {};
__export(focusable_exports, {
  Focusable: () => Focusable,
  getFocusableElements: () => getFocusableElements
});
module.exports = __toCommonJS(focusable_exports);
function getFocusableElements(container = document.body) {
  return {
    /**
     * Gets the number of keyboard focusable elements.
     * This is useful for checking if there are any focusable elements
     * before attempting to focus them.
     * 
     * @returns {number} The number of keyboard focusable elements
     */
    get length() {
      return this.keyboard.length;
    },
    /**
     * Gets all focusable elements within the container that are visible and enabled.
     * This includes elements that can be focused via mouse or programmatically,
     * but may not be keyboard focusable.
     * 
     * Elements are considered focusable if they:
     * - Are not disabled
     * - Are not hidden
     * - Are not display:none
     * 
     * @returns {HTMLElement[]} Array of focusable elements that are visible and enabled
     */
    get all() {
      const elements = Array.from(
        container.querySelectorAll(
          `a, 
            button, 
            input, 
            textarea, 
            select, 
            details, 
            iframe, 
            embed, 
            object, 
            summary, 
            dialog, 
            audio[controls], 
            video[controls], 
            [contenteditable], 
            [tabindex]
          `
        )
      );
      return elements.filter((el) => {
        if (el.hasAttribute("disabled")) return false;
        if (el.hasAttribute("hidden")) return false;
        if (window.getComputedStyle(el).display === "none") return false;
        return true;
      });
    },
    /**
     * Gets elements that can be focused via keyboard (tabIndex > -1).
     * These are the elements that will be focused when using the Tab key.
     * 
     * @returns {HTMLElement[]} Array of keyboard focusable elements
     */
    get keyboard() {
      return this.all.filter((el) => el instanceof HTMLElement && el.tabIndex > -1);
    },
    /**
     * Alias for keyboard focusable elements.
     * This is provided for semantic clarity in certain contexts.
     * 
     * @returns {HTMLElement[]} Array of keyboard focusable elements
     */
    get keyboardOnly() {
      return this.keyboard;
    },
    /**
     * Gets the first keyboard focusable element.
     * This is useful for setting initial focus when opening a modal or dialog.
     * 
     * @returns {HTMLElement|null} The first focusable element or null if none exist
     */
    get first() {
      return this.keyboard[0] || null;
    },
    /**
     * Gets the last keyboard focusable element.
     * This is useful for focus trapping when using Shift+Tab.
     * 
     * @returns {HTMLElement|null} The last focusable element or null if none exist
     */
    get last() {
      return this.keyboard[this.length - 1] || null;
    },
    /**
     * Gets the next focusable element after the given index.
     * This is useful for implementing custom tab order or focus management.
     * 
     * @param {number} index - The current element's index in the keyboard focusable array
     * @returns {HTMLElement|null} The next focusable element or null if none exists
     */
    next(index) {
      return this.keyboard[index + 1] || null;
    },
    /**
     * Gets the previous focusable element before the given index.
     * This is useful for implementing custom tab order or focus management.
     * 
     * @param {number} index - The current element's index in the keyboard focusable array
     * @returns {HTMLElement|null} The previous focusable element or null if none exists
     */
    prev(index) {
      return this.keyboard[index - 1] || null;
    }
  };
}
function Focusable(container = document.body) {
  return getFocusableElements(container);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Focusable,
  getFocusableElements
});
