// @splendidlabz/utils/dist/cjs/dom/font-size.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var font_size_exports = {};
__export(font_size_exports, {
  em: () => em,
  getUnit: () => getUnit,
  lh: () => lh,
  rem: () => rem,
  rlh: () => rlh,
  toPx: () => toPx
});
module.exports = __toCommonJS(font_size_exports);
var import_numbers = require("../lib/numbers/index.js");
function getUnit(value) {
  const [, unit] = (0, import_numbers.splitUnit)(value);
  return unit;
}
function rem(multiple = 1) {
  const value = parseFloat(getComputedStyle(document.body)["font-size"]);
  return Math.round(value * multiple);
}
function em(element = document.body, multiple = 1) {
  const value = parseFloat(getComputedStyle(element)["font-size"]);
  return Math.round(value * multiple);
}
function rlh(multiple = 1) {
  const lineHeight = parseFloat(getComputedStyle(document.body)["line-height"]);
  return Math.round(lineHeight * multiple);
}
function lh(element = document.body, multiple = 1) {
  const lineHeight = getComputedStyle(element)["line-height"];
  const value = parseFloat(lineHeight);
  return Math.round(value * multiple);
}
function toPx(value, element = null) {
  const [numeric, unit] = (0, import_numbers.splitUnit)(value);
  let finalValue = 0;
  if (unit === "rem") finalValue = rem(numeric);
  if (unit === "em") finalValue = em(element, numeric);
  if (unit === "lh") finalValue = lh(element, numeric);
  if (unit === "px") finalValue = numeric;
  if (!unit) finalValue = numeric;
  return finalValue;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  em,
  getUnit,
  lh,
  rem,
  rlh,
  toPx
});
