// @splendidlabz/utils/dist/cjs/dom/get-element.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_element_exports = {};
__export(get_element_exports, {
  getAncestorWithSiblings: () => getAncestorWithSiblings,
  getChildrenElements: () => getChildrenElements,
  getElement: () => getElement,
  getNodeType: () => getNodeType,
  getParentElement: () => getParentElement,
  getSelfIndex: () => getSelfIndex,
  getSiblingElements: () => getSiblingElements,
  isAncestor: () => isAncestor
});
module.exports = __toCommonJS(get_element_exports);
const astroNodes = ["ASTRO-SLOT", "ASTRO-ISLAND"];
function getNodeType(node) {
  if (node instanceof Element) return "element";
  if (node instanceof NodeList) return "nodelist";
  if (Array.isArray(node)) return "array";
}
function getElement(selector) {
  if (!selector) return;
  if (selector instanceof Element) return selector;
  return document.querySelector(selector);
}
function getChildrenElements(node) {
  let children = node.children;
  if (children[0]?.nodeName === "ASTRO-SLOT") children = children[0].children;
  return Array.from(children);
}
function getParentElement(element) {
  const parent = element.parentElement;
  if (astroNodes.includes(parent.nodeName)) return getParentElement(parent);
  return parent;
}
function getSiblingElements(element) {
  return Array.from(element.parentElement.children).filter(
    (child) => child !== element
  );
}
function getSelfIndex(element) {
  return Array.from(element.parentElement.children).indexOf(element);
}
function isAncestor(ancestor, element, { searchLevels = 0 } = {}) {
  if (ancestor === element) return true;
  let parent = element.parentElement;
  if (searchLevels === 0) {
    while (parent) {
      if (ancestor === parent) return true;
      parent = parent.parentElement;
    }
  } else {
    for (let i = 0; i < searchLevels; i++) {
      if (ancestor === parent) return true;
      parent = parent.parentElement;
    }
  }
  return false;
}
function getAncestorWithSiblings(element, { limit = false } = {}) {
  const ancestor = element.parentElement;
  const siblings = getSiblingElements(ancestor);
  if (siblings.length) return siblings;
  if (!ancestor) return null;
  if (limit === false) return getAncestorWithSiblings(ancestor);
  if (typeof limit === "number" && limit > 0) {
    return getAncestorWithSiblings(ancestor, { limit: limit - 1 });
  }
  if (limit === 0) return null;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getAncestorWithSiblings,
  getChildrenElements,
  getElement,
  getNodeType,
  getParentElement,
  getSelfIndex,
  getSiblingElements,
  isAncestor
});
