// @splendidlabz/utils/dist/cjs/dom/pkce.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pkce_exports = {};
__export(pkce_exports, {
  PKCE: () => PKCE
});
module.exports = __toCommonJS(pkce_exports);
var import_random_string = require("./random-string.js");
async function PKCE() {
  const codeVerifier = await (0, import_random_string.randomString)();
  const codeChallenge = await getCodeChallenge(codeVerifier);
  return {
    state: await (0, import_random_string.randomString)(),
    code_verifier: codeVerifier,
    code_challenge: codeChallenge,
    code_challenge_method: "S256"
  };
}
async function getCodeChallenge(verifier) {
  const encoder = new TextEncoder();
  const data = encoder.encode(verifier);
  const hashed = await window.crypto.subtle.digest("SHA-256", data);
  return window.btoa(String.fromCharCode.apply(null, new Uint8Array(hashed))).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PKCE
});
