// @splendidlabz/utils/src/lib/date/time.js
import { splitUnit } from '../numbers/index.js'

/**
 * Convert a time string to milliseconds
 * @param {string} arg - The time string to convert
 * @returns {number}
 */
export function ms(arg) {
  const [value, unit] = splitUnit(arg)
  if (unit === 'd') return getTimeInMs(value, 'days')
  if (unit === 'day') return getTimeInMs(value, 'days')
  if (unit === 'days') return getTimeInMs(value, 'days')
  if (unit === 'h') return getTimeInMs(value, 'hours')
  if (unit === 'hr') return getTimeInMs(value, 'hours')
  if (unit === 'hour') return getTimeInMs(value, 'hours')
  if (unit === 'hours') return getTimeInMs(value, 'hours')
  if (unit === 'm') return getTimeInMs(value, 'minutes')
  if (unit === 'min') return getTimeInMs(value, 'minutes')
  if (unit === 'mins') return getTimeInMs(value, 'minutes')
  if (unit === 'minute') return getTimeInMs(value, 'minutes')
  if (unit === 'minutes') return getTimeInMs(value, 'minutes')
  if (unit === 's') return getTimeInMs(value, 'seconds')
  if (unit === 'sec') return getTimeInMs(value, 'seconds')
  if (unit === 'secs') return getTimeInMs(value, 'seconds')
  if (unit === 'second') return getTimeInMs(value, 'seconds')
  if (unit === 'seconds') return getTimeInMs(value, 'seconds')

  // The rest are for milliseconds
  return value
}

export function getTimeInMs(value, unit = 'ms') {
  switch (unit) {
    case 'ms':
      return value
    case 'seconds':
      return value * 1000
    case 'minutes':
      return value * 1000 * 60
    case 'hours':
      return value * 1000 * 60 * 60
    case 'days':
      return value * 1000 * 60 * 60 * 24
    case 'weeks':
      return value * 1000 * 60 * 60 * 24 * 7
  }
}

export function toWeeks(ms) {
  return ms / 1000 / 60 / 60 / 24 / 7
}

export function toDays(ms) {
  return ms / 1000 / 60 / 60 / 24
}

export function toHours(ms) {
  return ms / 1000 / 60 / 60
}

export function toMinutes(ms) {
  return ms / 1000 / 60
}

export function toSeconds(ms) {
  return ms / 1000
}

// Get the total number of weeks, days, hours, minutes, and seconds
export function toReadableTime(ms) {
  const weekMs = 1000 * 60 * 60 * 24 * 7
  const dayMs = 1000 * 60 * 60 * 24
  const hourMs = 1000 * 60 * 60
  const minuteMs = 1000 * 60

  const weeks = Math.floor(ms / weekMs)
  const days = Math.floor((ms % weekMs) / dayMs)
  const hours = Math.floor((ms % dayMs) / hourMs)
  const minutes = Math.floor((ms % hourMs) / minuteMs)
  const seconds = Math.floor((ms % minuteMs) / 1000)

  return { weeks, days, hours, minutes, seconds }
}
