// @splendidlabz/utils/dist/esm/lib/objects/mix/mix.js
const DEFAULT_OPTIONS = { array: "replace" };
const mix = createMix();
const concatMix = createMix({ array: "concat" });
function createMix(userOptions = {}) {
  const options = { ...DEFAULT_OPTIONS, ...userOptions };
  function _mix(...sources) {
    const result = {};
    for (const source of sources) {
      if (source == null) continue;
      merge(result, source, options);
    }
    return result;
  }
  return _mix;
}
function merge(output, input, options) {
  const props = Object.keys(input);
  for (const prop of props) {
    if (prop === "__proto__") continue;
    const descriptor = Object.getOwnPropertyDescriptor(input, prop);
    const value = descriptor.value;
    if (value) descriptor.value = cloneDescriptorValue(value);
    if (!output[prop]) {
      Object.defineProperty(output, prop, descriptor);
      continue;
    }
    if (typeof output[prop] !== "object") {
      Object.defineProperty(output, prop, descriptor);
      continue;
    }
    if (descriptor.value == null || typeof descriptor.value !== "object") {
      Object.defineProperty(output, prop, descriptor);
      continue;
    }
    if (objectType(descriptor.value) === "[object Array]") {
      output[prop] = options.array === "replace" ? descriptor.value.slice() : output[prop].concat(descriptor.value);
      continue;
    }
    merge(output[prop], descriptor.value, options);
  }
}
function cloneDescriptorValue(value) {
  if (objectType(value) === "[object Array]") {
    const array = [];
    for (let v of value) {
      v = cloneDescriptorValue(v);
      array.push(v);
    }
    return array;
  }
  if (objectType(value) === "[object Object]") {
    const obj = {};
    const props = Object.keys(value);
    for (const prop of props) {
      const descriptor = Object.getOwnPropertyDescriptor(value, prop);
      if (descriptor.value)
        descriptor.value = cloneDescriptorValue(descriptor.value);
      Object.defineProperty(obj, prop, descriptor);
    }
    return obj;
  }
  if (objectType(value) === "[object Date]") {
    return new Date(value.getTime());
  }
  if (objectType(value) === "[object Map]") {
    const map = /* @__PURE__ */ new Map();
    for (const entry of value) {
      map.set(entry[0], cloneDescriptorValue(entry[1]));
    }
    return map;
  }
  if (objectType(value) === "[object Set]") {
    const set = /* @__PURE__ */ new Set();
    for (const entry of value.entries()) {
      set.add(cloneDescriptorValue(entry[0]));
    }
    return set;
  }
  return value;
}
function objectType(value) {
  return Object.prototype.toString.call(value);
}
export {
  concatMix,
  createMix,
  mix
};
