// @splendidlabz/utils/src/dom/actions/prefer-horizontal-scroll.js
import { omitEmpty } from '../../lib/objects/omit-empty.js'
import { getCSSVar, setCSSVar } from '../css-vars.js'
import { addListeners, removeListeners } from '../events.js'

const DEFAULT_OPTIONS = {
  scrollSnapDelay: 1000,
}

export function preferHorizontalScroll(node, props = {}) {
  if (!node.classList.contains('scrollable-prefer-horizontal')) return

  const state = {
    origSnapType: null,
    currentSnapType: null,
    snapTimeoutID: null,
  }

  const options = {
    ...DEFAULT_OPTIONS,
    ...omitEmpty({
      scrollSnapDelay: getCSSVar(node, '--scroll-snap-delay'),
    }),
    ...omitEmpty(props),
  }

  const listeners = [
    {
      event: 'wheel',
      node,
      handler: handleWheel,
    },

    {
      event: 'pointerenter',
      node,
      handler: handlePointerEnter,
    },

    {
      event: 'pointerleave',
      node,
      handler: handlePointerLeave,
    },
  ]

  addListeners(listeners)

  function handleWheel(event) {
    const direction = event.deltaY > 0 ? 'down' : 'up'
    const updatedDelta = event.deltaY

    // Only handle mouse wheel
    // Trackpad users can scroll horizontally themselves
    if (isTrackpad(event)) return false

    setCSSVar(node, '--snap-type', 'none')

    const scrolledAmount = node.scrollLeft
    const scrollWidth = node.scrollWidth
    const clientWidth = node.clientWidth
    const scrollableAmount = scrollWidth - clientWidth // Maximum scrollable width to the right

    // Hijack downwards scroll
    if (direction === 'down' && scrolledAmount < scrollableAmount) {
      node.scrollLeft += updatedDelta
      event.preventDefault()
    }

    // Hijack upwards scroll
    if (direction === 'up' && scrolledAmount > 0) {
      node.scrollLeft += updatedDelta
      event.preventDefault()
    }

    // Restore snap type after a delay
    const { snapTimeoutID } = state
    if (snapTimeoutID) clearTimeout(snapTimeoutID)
    state.snapTimeoutID = setTimeout(_ => {
      restoreHijackedState(node, state)
    }, options.scrollSnapDelay)
  }

  function handlePointerEnter(event) {
    state.origSnapType = getCSSVar(node, '--snap-type') || 'proximity'
  }

  function handlePointerLeave(event) {
    restoreHijackedState(node, state)
  }

  return {
    destroy() {
      removeListeners(listeners)
    },
  }
}

function restoreHijackedState(node, state) {
  const scrollLeft = node.scrollLeft

  // Get the final scroll position
  setCSSVar(node, '--snap-type', state.origSnapType)
  const finalScrollLeft = node.scrollLeft

  // Scroll back to original position before replacing snap
  setCSSVar(node, '--snap-type', 'none')
  node.scrollTo({ left: scrollLeft, behavior: 'instant' })

  // Scroll back to the final position smoothly
  node.scrollTo({ left: finalScrollLeft, behavior: 'smooth' })

  setTimeout(_ => {
    setCSSVar(node, '--snap-type', state.origSnapType)
  }, 250)
}

// https://stackoverflow.com/questions/10744645/detect-touchpad-vs-mouse-in-javascript/62415754#62415754
function isTrackpad(event) {
  // if (event.deltaMode === 0) return true // Deltamode no longer works in Chrome
  if (event.wheelDeltaY === event.deltaY * -3) return true
}
