// @splendidlabz/utils/dist/cjs/dom/actions/prefer-horizontal-scroll.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var prefer_horizontal_scroll_exports = {};
__export(prefer_horizontal_scroll_exports, {
  preferHorizontalScroll: () => preferHorizontalScroll
});
module.exports = __toCommonJS(prefer_horizontal_scroll_exports);
var import_omit_empty = require("../../lib/objects/omit-empty.js");
var import_css_vars = require("../css-vars.js");
var import_events = require("../events.js");
const DEFAULT_OPTIONS = {
  scrollSnapDelay: 1e3
};
function preferHorizontalScroll(node, props = {}) {
  if (!node.classList.contains("scrollable-prefer-horizontal")) return;
  const state = {
    origSnapType: null,
    currentSnapType: null,
    snapTimeoutID: null
  };
  const options = {
    ...DEFAULT_OPTIONS,
    ...(0, import_omit_empty.omitEmpty)({
      scrollSnapDelay: (0, import_css_vars.getCSSVar)(node, "--scroll-snap-delay")
    }),
    ...(0, import_omit_empty.omitEmpty)(props)
  };
  const listeners = [
    {
      event: "wheel",
      node,
      handler: handleWheel
    },
    {
      event: "pointerenter",
      node,
      handler: handlePointerEnter
    },
    {
      event: "pointerleave",
      node,
      handler: handlePointerLeave
    }
  ];
  (0, import_events.addListeners)(listeners);
  function handleWheel(event) {
    const direction = event.deltaY > 0 ? "down" : "up";
    const updatedDelta = event.deltaY;
    if (isTrackpad(event)) return false;
    (0, import_css_vars.setCSSVar)(node, "--snap-type", "none");
    const scrolledAmount = node.scrollLeft;
    const scrollWidth = node.scrollWidth;
    const clientWidth = node.clientWidth;
    const scrollableAmount = scrollWidth - clientWidth;
    if (direction === "down" && scrolledAmount < scrollableAmount) {
      node.scrollLeft += updatedDelta;
      event.preventDefault();
    }
    if (direction === "up" && scrolledAmount > 0) {
      node.scrollLeft += updatedDelta;
      event.preventDefault();
    }
    const { snapTimeoutID } = state;
    if (snapTimeoutID) clearTimeout(snapTimeoutID);
    state.snapTimeoutID = setTimeout((_) => {
      restoreHijackedState(node, state);
    }, options.scrollSnapDelay);
  }
  function handlePointerEnter(event) {
    state.origSnapType = (0, import_css_vars.getCSSVar)(node, "--snap-type") || "proximity";
  }
  function handlePointerLeave(event) {
    restoreHijackedState(node, state);
  }
  return {
    destroy() {
      (0, import_events.removeListeners)(listeners);
    }
  };
}
function restoreHijackedState(node, state) {
  const scrollLeft = node.scrollLeft;
  (0, import_css_vars.setCSSVar)(node, "--snap-type", state.origSnapType);
  const finalScrollLeft = node.scrollLeft;
  (0, import_css_vars.setCSSVar)(node, "--snap-type", "none");
  node.scrollTo({ left: scrollLeft, behavior: "instant" });
  node.scrollTo({ left: finalScrollLeft, behavior: "smooth" });
  setTimeout((_) => {
    (0, import_css_vars.setCSSVar)(node, "--snap-type", state.origSnapType);
  }, 250);
}
function isTrackpad(event) {
  if (event.wheelDeltaY === event.deltaY * -3) return true;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  preferHorizontalScroll
});
