// @splendidlabz/utils/dist/cjs/dom/actions/sticky.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sticky_exports = {};
__export(sticky_exports, {
  sticky: () => sticky
});
module.exports = __toCommonJS(sticky_exports);
var import_bounding_box = require("../bounding-box.js");
var import_resize_observer = require("../observers/resize-observer.js");
var import_scroll_container = require("../ui/scroll-container.js");
function sticky(node, props = {}) {
  const scrollContainer = (0, import_scroll_container.findScrollContainer)(node);
  let stickyTBLR = getComputedTBLR(node);
  const obs = (0, import_resize_observer.resizeObserver)(node, {
    callback: ({ entry }) => {
      stickyTBLR = getComputedTBLR(node);
      isSticking();
      scrollContainer === document.documentElement ? window.addEventListener("scroll", scrollEventListener) : scrollContainer.addEventListener("scroll", scrollEventListener);
    }
  });
  function scrollEventListener(event) {
    isSticking();
  }
  function isSticking() {
    const { top, bottom, left, right } = stickyTBLR;
    const relativeRect = (0, import_bounding_box.boundingBoxRelativeToAncestor)({
      element: node,
      ancestorElement: scrollContainer
    });
    if (!isNaN(top)) {
      if (relativeRect.top - top <= 1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(bottom)) {
      if (relativeRect.bottom + bottom >= -1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(left)) {
      if (relativeRect.left - left <= 1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(right)) {
      if (relativeRect.right + right >= -1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
  }
  return {
    destroy() {
      obs.disconnect();
      scrollContainer.removeEventListener("scroll", scrollEventListener);
    }
  };
}
function getComputedTBLR(element) {
  const style = getComputedStyle(element);
  return {
    top: parseFloat(style.top),
    bottom: parseFloat(style.bottom),
    left: parseFloat(style.left),
    right: parseFloat(style.right)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  sticky
});
