// @splendidlabz/utils/src/dom/observers/intersection-observer.js
import { dispatchEvent } from '../events.js'
import { useObserverMethodOnTarget } from './observer.js'

// See https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver#options for options
export function intersectionObserver(target, options = {}) {
  const { callback, ...opts } = options
  const observer = new IntersectionObserver(observerFn, opts)

  useObserverMethodOnTarget(target, observer, 'observe')

  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer })
      else dispatchEvent(target, 'intersect', { entry, entries, observer })
    }
  }

  return {
    observer,
    observe(target) {
      useObserverMethodOnTarget(target, observer, 'observe', options)
    },

    unobserve(target) {
      useObserverMethodOnTarget(target, observer, 'unobserve')
    },

    takeRecords() {
      return observer.takeRecords()
    },

    disconnect() {
      // Take records before disconnecting.
      const records = observer.takeRecords()
      observer.disconnect()
      if (records.length > 0) observerFn(records)
    },

    destroy() {
      observer.disconnect()
    },
  }
}
