// @splendidlabz/utils/src/dom/observers/mutation-observer.js
/* eslint-env browser */
import { dispatchEvent } from '../events.js'
import { useObserverMethodOnTarget } from './observer.js'

// https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe

// These are simply options that will often be used. Check the docs for other options. Adding them here simply because it's easier to refer them here than reading the docs.
const defaultOptions = {
  attributes: false,
  childList: false,
  subtree: false,
}

export function mutationObserver(target, options) {
  options = { ...defaultOptions, ...options }
  const { callback, ...opts } = options
  const observer = new MutationObserver(observerFn)

  if (target === window) target = document.body
  useObserverMethodOnTarget(target, observer, 'observe', opts)

  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer })
      else dispatchEvent(target, 'mutate', { entry, entries, observer })
    }
  }

  return {
    observer,
    observe(target) {
      useObserverMethodOnTarget(target, observer, 'observe', options)
    },

    takeRecords() {
      return observer.takeRecords()
    },

    disconnect() {
      // Take records before disconnecting.
      // https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/takeRecords
      const records = observer.takeRecords()
      observer.disconnect()
      if (records.length > 0) observerFn(records)
    },

    destroy() {
      observer.disconnect()
    },
  }
}
