// @splendidlabz/utils/src/dom/observers/resize-observer.js
/* eslint-env browser */
import { dispatchEvent } from '../events.js'
import { useObserverMethodOnTarget } from './observer.js'

/**
 * Creates and manages a ResizeObserver instance to monitor size changes of a target element.
 *
 * @param {Element|Window|Document|NodeList|Element[]} target - The element(s) to observe.
 *   - If window or document is provided, document.documentElement will be observed instead.
 *   - If NodeList or Array of elements is provided, all elements will be observed.
 * @param {Object} options - Configuration options for the resize observer
 * @param {boolean} [options.observe=true] - Whether to start observing immediately. If false, the observer won't be created.
 * @param {Function} [options.callback] - Optional callback function that will be called when resize changes are detected.
 *                                          If not provided, a 'resize-obs' event will be dispatched on the target.
 * @param {Object} [options.observerOptions] - Additional options to pass to ResizeObserver.observe()
 *
 * @returns {Object} An object with methods to control the observer:
 *   - observe(target, options): Start observing a new target element
 *   - unobserve(target): Stop observing a target element
 *   - disconnect(): Disconnect the observer and stop all observations
 *   - destroy(): Alias for disconnect()
 *
 * @example
 * // Basic usage with callback
 * resizeObserver(element, {
 *   callback: ({ entry, entries, observer }) => {
 *     console.log('Element resized:', entry.contentRect);
 *   }
 * });
 *
 * @example
 * // Usage with event listener
 * resizeObserver(element);
 * element.addEventListener('resize-obs', ({ detail }) => {
 *   console.log('Element resized:', detail.entry.contentRect);
 * });
 */
export function resizeObserver(target, options) {
  const { callback, ...opts } = options
  const observer = new ResizeObserver(observerFn)

  if (target === window || target === document)
    target = document.documentElement
  useObserverMethodOnTarget(target, observer, 'observe', opts)

  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer })
      else
        dispatchEvent(entry.target, 'resize-obs', { entry, entries, observer })
    }
  }

  return {
    observe(target) {
      useObserverMethodOnTarget(target, observer, 'observe', options)
    },

    unobserve(target) {
      useObserverMethodOnTarget(target, observer, 'unobserve')
    },

    disconnect: _ => observer.disconnect(),
    destroy: _ => observer.disconnect(),
  }
}
