// @splendidlabz/utils/src/dom/observers/scroll-observer.js
/* eslint-env browser */
import { resizeObserver } from './resize-observer.js'

const defaultOptions = {
  threshold: 0, // Float between 0 to 1.
  tolerance: 0.1, // Float between 0 to 1. Tolerance for event firing
  once: false, // Only fire threshold callback once
}

/**
 * @typedef {Object} ScrollObserverOptions
 * @property {number} [threshold] - Float between 0 to 1. When to trigger callback (0.75 = 75% down page)
 * @property {number} [tolerance] - Float between 0 to 1. Tolerance zone around threshold
 * @property {boolean} [once] - Only fire threshold callback once (default: false)
 * @property {Function} [callback] - Called on every scroll with scrollPercent
 * @property {Function} [onScrollDown] - Called when scrolling down
 * @property {Function} [onScrollUp] - Called when scrolling up
 * @property {Function} [onEnterThreshold] - Called when entering threshold zone
 */

/**
 * @param {Element | Document | Window} node
 * @returns {Element}
 */
function getScrollElement(node) {
  if (
    node === document ||
    node === window ||
    node === document.documentElement
  ) {
    return document.documentElement
  }
  return /** @type {Element} */ (node)
}

/**
 * Scroll Observer - Optimized for performance
 * @param {Element | Document | Window} node - The element to observe scroll events on
 * @param {ScrollObserverOptions} [options] - Configuration options
 */
export function scrollObserver(node, options = {}) {
  const { callback, onScrollDown, onScrollUp, onEnterThreshold, ...userOpts } =
    options
  const opts = { ...defaultOptions, ...userOpts }
  const { threshold, tolerance, once } = opts

  let prevScrollDirection = null
  let prevScrollTop = 0
  let prevScrollPercent = 0
  let thresholdFired = false
  let rafId = null

  // Determine scroll context once at initialization
  const isDocumentScroll =
    node === document || node === window || node === document.documentElement
  const scrollElement = getScrollElement(node)
  const eventTarget = isDocumentScroll ? window : node

  // Cache DOM references and expensive calculations
  let cachedScrollHeight = 0
  let cachedClientHeight = 0

  // Initialize cache and start observing for changes
  updateCache()
  const cacheObserver = resizeObserver(scrollElement, {
    callback: updateCache,
  })

  eventTarget.addEventListener('scroll', throttledObserve, { passive: true })

  function throttledObserve() {
    rafId = requestAnimationFrame(observe)
  }

  function updateCache() {
    cachedScrollHeight = scrollElement.scrollHeight
    cachedClientHeight = scrollElement.clientHeight
  }

  function observe() {
    const scrollTop = scrollElement.scrollTop

    // Skip if scroll position hasn't changed meaningfully
    if (Math.abs(scrollTop - prevScrollTop) < 1) return

    const scrollDirection = scrollTop > prevScrollTop ? 'down' : 'up'
    const maxScroll = Math.max(1, cachedScrollHeight - cachedClientHeight)
    const scrollPercent = Math.min(1, Math.max(0, scrollTop / maxScroll))

    // Check threshold crossing (e.g., 75% ± tolerance)
    const thresholdMin = threshold - tolerance / 2
    const thresholdMax = threshold + tolerance / 2
    const wasInThreshold =
      prevScrollPercent >= thresholdMin && prevScrollPercent <= thresholdMax
    const isInThreshold =
      scrollPercent >= thresholdMin && scrollPercent <= thresholdMax
    const hasEnteredThreshold =
      !wasInThreshold && isInThreshold && (!once || !thresholdFired)

    if (hasEnteredThreshold && once) {
      thresholdFired = true
    }

    // Prepare common data for all callbacks
    const callbackData = {
      scrollTop,
      scrollDirection,
      scrollPercent,
      directionChanged: scrollDirection !== prevScrollDirection,
      hasEnteredThreshold,
      isInThreshold,
    }

    // Call specific callbacks
    if (typeof callback === 'function') {
      callback(callbackData)
    }

    // Only fire direction callbacks when direction actually changes
    if (scrollDirection !== prevScrollDirection) {
      if (scrollDirection === 'down' && typeof onScrollDown === 'function') {
        onScrollDown(callbackData)
      }
      if (scrollDirection === 'up' && typeof onScrollUp === 'function') {
        onScrollUp(callbackData)
      }
    }

    // Fire threshold callback when entering threshold zone
    if (hasEnteredThreshold && typeof onEnterThreshold === 'function') {
      onEnterThreshold(callbackData)
    }

    prevScrollTop = scrollTop
    prevScrollPercent = scrollPercent
    prevScrollDirection = scrollDirection
  }

  return {
    destroy() {
      eventTarget.removeEventListener('scroll', throttledObserve)
      if (rafId) cancelAnimationFrame(rafId)
      cacheObserver.destroy()
    },
  }
}
