// @splendidlabz/utils/src/lib/auth/route-manager.js
/**
  // Usage ... Probably should use either protected or bypass, not both.
  // Maybe ... glob would be a better for this
  @example 
  const config = {
    routeConfig: {
      startsWith: ['/dashboard']
    }
    bypassRoutes: {
      is: ['/'],
      startsWith: ['/fonts', '/auth', '/favicon', '/images'],
      includes: ['/meh'],
    }
  }
  const {isProtectedRoute, isBypassRoute} = AuthManager(config)
  if (!isProtectedRoute(context.url.pathname)) return next() // Handle only protected routes
  if (isBypassRoute(context.url.pathname)) return next() // Bypass all non protected routes
 */

export function RouteManager(config) {
  const { protectedRoutes, bypassRoutes } = config
  return {
    isProtectedRoute(pathname) {
      return isMatch(protectedRoutes, pathname)
    },
    isBypassRoute(pathname) {
      return isMatch(bypassRoutes, pathname)
    },
  }
}

// Alias for RouteManager
export function AuthManager(config) {
  return RouteManager(config)
}

function isMatch(routeConfig, pathname) {
  const { is, startsWith, includes } = routeConfig
  const isExactMatch = is && is.some(u => pathname === u)
  const startsWithMatch =
    startsWith && startsWith.some(u => pathname.startsWith(u))
  const includesMatch = includes && includes.some(u => pathname.includes(u))

  return isExactMatch || startsWithMatch || includesMatch
}
