// @splendidlabz/utils/src/lib/strings/markdown.js
import markdownit from 'markdown-it'
import anchor from 'markdown-it-anchor'

const md = markdownit({
  html: true,
  linkify: true,
  typographer: true,
})

md.use(anchor, {
  permalink: false,
  slugify: s =>
    s
      .toLowerCase()
      .replace(/\s+/g, '-')
      .replace(/[^\w-]/g, ''),
})

/**
 * Converts markdown content to HTML with whitespace handling
 * @param {string} content - Markdown content to convert
 * @param {Object} [options={}] - Conversion options
 * @param {boolean} [options.inline=false] - Whether content should be treated as inline markdown
 * @returns {string} Converted HTML string
 */
export function markdown(content, options = {}) {
  const defaultOptions = { inline: false }
  const opts = { ...defaultOptions, ...options }

  const { content: treated } = treatMarkdownWhitespace(content, opts)

  if (opts.inline) return md.renderInline(treated)
  return md.render(treated)
}

/**
 * Handles whitespace in markdown content to ensure proper rendering
 * @param {string} content - Markdown content to process
 * @param {Object} [options={}] - Processing options
 * @param {boolean} [options.inline=false] - Whether content is inline markdown
 * @returns {{ content: string, inline: boolean }} Processed content and inline status
 */
export function treatMarkdownWhitespace(content, { inline } = {}) {
  if (!content) return { content: '', inline }
  const lines = content.split('\n')

  // Check for inline content
  if (lines[0] !== '' && inline) return { content, inline: true }

  // If it's not inline content, we need to strip indentation
  // For this, we need to check for the first new line with leading spaces
  const firstLineWithLeadingSpacesIndex = lines.findIndex(line =>
    line.match(/^\s+/),
  )
  const firstLineWithLeadingSpaces = lines[firstLineWithLeadingSpacesIndex]
  const indentation =
    firstLineWithLeadingSpaces?.match(/^\s+/)?.[0]?.length ?? 0

  // Once the indentation is found, we strip the indentation from each line. This treats the Markdown whitespace and prevents it from being rendered as code blocks.
  const retContent = lines
    .map((line, index) => {
      if (index < firstLineWithLeadingSpacesIndex) return line + '\n'
      return line.slice(indentation) + '\n'
    })
    .join('')

  return {
    content: retContent,
    inline,
  }
}
