// @splendidlabz/utils/dist/esm/lib/date/time.js
import { splitUnit } from "../numbers/index.js";
function ms(arg) {
  const [value, unit] = splitUnit(arg);
  if (unit === "d") return getTimeInMs(value, "days");
  if (unit === "day") return getTimeInMs(value, "days");
  if (unit === "days") return getTimeInMs(value, "days");
  if (unit === "h") return getTimeInMs(value, "hours");
  if (unit === "hr") return getTimeInMs(value, "hours");
  if (unit === "hour") return getTimeInMs(value, "hours");
  if (unit === "hours") return getTimeInMs(value, "hours");
  if (unit === "m") return getTimeInMs(value, "minutes");
  if (unit === "min") return getTimeInMs(value, "minutes");
  if (unit === "mins") return getTimeInMs(value, "minutes");
  if (unit === "minute") return getTimeInMs(value, "minutes");
  if (unit === "minutes") return getTimeInMs(value, "minutes");
  if (unit === "s") return getTimeInMs(value, "seconds");
  if (unit === "sec") return getTimeInMs(value, "seconds");
  if (unit === "secs") return getTimeInMs(value, "seconds");
  if (unit === "second") return getTimeInMs(value, "seconds");
  if (unit === "seconds") return getTimeInMs(value, "seconds");
  return value;
}
function getTimeInMs(value, unit = "ms") {
  switch (unit) {
    case "ms":
      return value;
    case "seconds":
      return value * 1e3;
    case "minutes":
      return value * 1e3 * 60;
    case "hours":
      return value * 1e3 * 60 * 60;
    case "days":
      return value * 1e3 * 60 * 60 * 24;
    case "weeks":
      return value * 1e3 * 60 * 60 * 24 * 7;
  }
}
function toWeeks(ms2) {
  return ms2 / 1e3 / 60 / 60 / 24 / 7;
}
function toDays(ms2) {
  return ms2 / 1e3 / 60 / 60 / 24;
}
function toHours(ms2) {
  return ms2 / 1e3 / 60 / 60;
}
function toMinutes(ms2) {
  return ms2 / 1e3 / 60;
}
function toSeconds(ms2) {
  return ms2 / 1e3;
}
function toReadableTime(ms2) {
  const weekMs = 1e3 * 60 * 60 * 24 * 7;
  const dayMs = 1e3 * 60 * 60 * 24;
  const hourMs = 1e3 * 60 * 60;
  const minuteMs = 1e3 * 60;
  const weeks = Math.floor(ms2 / weekMs);
  const days = Math.floor(ms2 % weekMs / dayMs);
  const hours = Math.floor(ms2 % dayMs / hourMs);
  const minutes = Math.floor(ms2 % hourMs / minuteMs);
  const seconds = Math.floor(ms2 % minuteMs / 1e3);
  return { weeks, days, hours, minutes, seconds };
}
export {
  getTimeInMs,
  ms,
  toDays,
  toHours,
  toMinutes,
  toReadableTime,
  toSeconds,
  toWeeks
};
