// @splendidlabz/utils/dist/esm/lib/strings/convert-case/convert-case.js
function toKebab(string) {
  if (!string) return "";
  return string.split("").map((letter, index) => {
    const previousLetter = string[index - 1] || "";
    const currentLetter = letter;
    if (isDigit(currentLetter) && !isDigit(previousLetter)) {
      return `-${currentLetter}`;
    }
    if (!isCaps(currentLetter)) return currentLetter;
    if (previousLetter === "") {
      return `${currentLetter.toLowerCase()}`;
    }
    if (isCaps(previousLetter)) {
      return `${currentLetter.toLowerCase()}`;
    }
    return `-${currentLetter.toLowerCase()}`;
  }).join("").trim().replace(/[-_\s]+/g, "-");
}
function toCamel(string) {
  return toKebab(string).split("-").map((word, index) => {
    if (index === 0) return word;
    return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
  }).join("");
}
function toPascal(string) {
  const camel = toCamel(string);
  return camel.slice(0, 1).toUpperCase() + camel.slice(1);
}
function toTitle(string) {
  return toKebab(string).split("-").map((word) => {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
  }).join(" ");
}
function toSlug(string) {
  return toKebab(string);
}
function toSentence(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.slice(0, 1).toUpperCase() + interim.slice(1);
}
function toUpper(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.toUpperCase();
}
function toLower(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.toLowerCase();
}
function isCaps(char) {
  return /\p{Lu}/u.test(char);
}
function isDigit(char) {
  return /[0-9]/.test(char);
}
export {
  toCamel,
  toKebab,
  toLower,
  toPascal,
  toSentence,
  toSlug,
  toTitle,
  toUpper
};
