// @splendidlabz/utils/dist/esm/lib/functions/functional.js
function curry(fn) {
  return function curried(...args) {
    if (args.length >= fn.length) {
      return fn.apply(this, args);
    } else {
      return function(...moreArgs) {
        return curried.apply(this, args.concat(moreArgs));
      };
    }
  };
}
function compose(...fns) {
  return function(value) {
    return fns.reduceRight((acc, fn) => fn(acc), value);
  };
}
function composeAsync(...fns) {
  return async function(value) {
    return fns.reduceRight(async (acc, fn) => {
      const result = await acc;
      return await fn(result);
    }, value);
  };
}
function pipe(...fns) {
  return function(value) {
    return fns.reduce((acc, fn) => fn(acc), value);
  };
}
function pipeAsync(...fns) {
  return async function(value) {
    return fns.reduce(async (acc, fn) => {
      const result = await acc;
      return await fn(result);
    }, value);
  };
}
function times(fn, n) {
  const result = [];
  for (let i = 0; i < n; i++) {
    result.push(fn(i));
  }
  return result;
}
export {
  compose,
  composeAsync,
  curry,
  pipe,
  pipeAsync,
  times
};
