// @splendidlabz/utils/src/lib/strings/convert-case/convert-case.js
// ========================
// Case Conversion Utilities
// Note: All case conversion functions require `toKebab` to work.
// ------------------------
export function toKebab(string) {
  if (!string) return ''
  return string
    .split('')
    .map((letter, index) => {
      const previousLetter = string[index - 1] || ''
      const currentLetter = letter

      if (isDigit(currentLetter) && !isDigit(previousLetter)) {
        return `-${currentLetter}`
      }

      if (!isCaps(currentLetter)) return currentLetter

      if (previousLetter === '') {
        return `${currentLetter.toLowerCase()}`
      }

      if (isCaps(previousLetter)) {
        return `${currentLetter.toLowerCase()}`
      }

      return `-${currentLetter.toLowerCase()}`
    })
    .join('')
    .trim()
    .replace(/[-_\s]+/g, '-')
}

export function toCamel(string) {
  return toKebab(string)
    .split('-')
    .map((word, index) => {
      if (index === 0) return word
      return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase()
    })
    .join('')
}

export function toPascal(string) {
  const camel = toCamel(string)
  return camel.slice(0, 1).toUpperCase() + camel.slice(1)
}

export function toTitle(string) {
  return toKebab(string)
    .split('-')
    .map(word => {
      return word.slice(0, 1).toUpperCase() + word.slice(1)
    })
    .join(' ')
}

export function toSlug(string) {
  return toKebab(string)
}

export function toSentence(string) {
  const interim = toKebab(string).replace(/-/g, ' ')
  return interim.slice(0, 1).toUpperCase() + interim.slice(1)
}

export function toUpper(string) {
  const interim = toKebab(string).replace(/-/g, ' ')
  return interim.toUpperCase()
}

export function toLower(string) {
  const interim = toKebab(string).replace(/-/g, ' ')
  return interim.toLowerCase()
}

// Checks whether character is Uppercase.
// Crude version. Checks only A-Z.
function isCaps(char) {
  return /\p{Lu}/u.test(char)
}

// Checks whether character is digit.
function isDigit(char) {
  return /[0-9]/.test(char)
}
