// @splendidlabz/utils/dist/esm/dom/actions/sticky.js
import { boundingBoxRelativeToAncestor } from "../bounding-box.js";
import { resizeObserver } from "../observers/resize-observer.js";
import { findScrollContainer } from "../ui/scroll-container.js";
function sticky(node, props = {}) {
  const scrollContainer = findScrollContainer(node);
  let stickyTBLR = getComputedTBLR(node);
  const obs = resizeObserver(node, {
    callback: ({ entry }) => {
      stickyTBLR = getComputedTBLR(node);
      isSticking();
      scrollContainer === document.documentElement ? window.addEventListener("scroll", scrollEventListener) : scrollContainer.addEventListener("scroll", scrollEventListener);
    }
  });
  function scrollEventListener(event) {
    isSticking();
  }
  function isSticking() {
    const { top, bottom, left, right } = stickyTBLR;
    const relativeRect = boundingBoxRelativeToAncestor({
      element: node,
      ancestorElement: scrollContainer
    });
    if (!isNaN(top)) {
      if (relativeRect.top - top <= 1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(bottom)) {
      if (relativeRect.bottom + bottom >= -1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(left)) {
      if (relativeRect.left - left <= 1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
    if (!isNaN(right)) {
      if (relativeRect.right + right >= -1) {
        node.classList.add("sticking");
      } else {
        node.classList.remove("sticking");
      }
    }
  }
  return {
    destroy() {
      obs.disconnect();
      scrollContainer.removeEventListener("scroll", scrollEventListener);
    }
  };
}
function getComputedTBLR(element) {
  const style = getComputedStyle(element);
  return {
    top: parseFloat(style.top),
    bottom: parseFloat(style.bottom),
    left: parseFloat(style.left),
    right: parseFloat(style.right)
  };
}
export {
  sticky
};
