// @splendidlabz/utils/dist/esm/dom/observers/intersection-observer.js
import { dispatchEvent } from "../events.js";
import { useObserverMethodOnTarget } from "./observer.js";
function intersectionObserver(target, options = {}) {
  const { callback, ...opts } = options;
  const observer = new IntersectionObserver(observerFn, opts);
  useObserverMethodOnTarget(target, observer, "observe");
  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer });
      else dispatchEvent(target, "intersect", { entry, entries, observer });
    }
  }
  return {
    observer,
    observe(target2) {
      useObserverMethodOnTarget(target2, observer, "observe", options);
    },
    unobserve(target2) {
      useObserverMethodOnTarget(target2, observer, "unobserve");
    },
    takeRecords() {
      return observer.takeRecords();
    },
    disconnect() {
      const records = observer.takeRecords();
      observer.disconnect();
      if (records.length > 0) observerFn(records);
    },
    destroy() {
      observer.disconnect();
    }
  };
}
export {
  intersectionObserver
};
