// @splendidlabz/utils/dist/esm/dom/observers/mutation-observer.js
import { dispatchEvent } from "../events.js";
import { useObserverMethodOnTarget } from "./observer.js";
const defaultOptions = {
  attributes: false,
  childList: false,
  subtree: false
};
function mutationObserver(target, options) {
  options = { ...defaultOptions, ...options };
  const { callback, ...opts } = options;
  const observer = new MutationObserver(observerFn);
  if (target === window) target = document.body;
  useObserverMethodOnTarget(target, observer, "observe", opts);
  function observerFn(entries) {
    for (const entry of entries) {
      if (callback) callback({ entry, entries, observer });
      else dispatchEvent(target, "mutate", { entry, entries, observer });
    }
  }
  return {
    observer,
    observe(target2) {
      useObserverMethodOnTarget(target2, observer, "observe", options);
    },
    takeRecords() {
      return observer.takeRecords();
    },
    disconnect() {
      const records = observer.takeRecords();
      observer.disconnect();
      if (records.length > 0) observerFn(records);
    },
    destroy() {
      observer.disconnect();
    }
  };
}
export {
  mutationObserver
};
