// @splendidlabz/utils/dist/esm/dom/observers/scroll-observer.js
import { resizeObserver } from "./resize-observer.js";
const defaultOptions = {
  threshold: 0,
  // Float between 0 to 1.
  tolerance: 0.1,
  // Float between 0 to 1. Tolerance for event firing
  once: false
  // Only fire threshold callback once
};
function getScrollElement(node) {
  if (node === document || node === window || node === document.documentElement) {
    return document.documentElement;
  }
  return (
    /** @type {Element} */
    node
  );
}
function scrollObserver(node, options = {}) {
  const { callback, onScrollDown, onScrollUp, onEnterThreshold, ...userOpts } = options;
  const opts = { ...defaultOptions, ...userOpts };
  const { threshold, tolerance, once } = opts;
  let prevScrollDirection = null;
  let prevScrollTop = 0;
  let prevScrollPercent = 0;
  let thresholdFired = false;
  let rafId = null;
  const isDocumentScroll = node === document || node === window || node === document.documentElement;
  const scrollElement = getScrollElement(node);
  const eventTarget = isDocumentScroll ? window : node;
  let cachedScrollHeight = 0;
  let cachedClientHeight = 0;
  updateCache();
  const cacheObserver = resizeObserver(scrollElement, {
    callback: updateCache
  });
  eventTarget.addEventListener("scroll", throttledObserve, { passive: true });
  function throttledObserve() {
    rafId = requestAnimationFrame(observe);
  }
  function updateCache() {
    cachedScrollHeight = scrollElement.scrollHeight;
    cachedClientHeight = scrollElement.clientHeight;
  }
  function observe() {
    const scrollTop = scrollElement.scrollTop;
    if (Math.abs(scrollTop - prevScrollTop) < 1) return;
    const scrollDirection = scrollTop > prevScrollTop ? "down" : "up";
    const maxScroll = Math.max(1, cachedScrollHeight - cachedClientHeight);
    const scrollPercent = Math.min(1, Math.max(0, scrollTop / maxScroll));
    const thresholdMin = threshold - tolerance / 2;
    const thresholdMax = threshold + tolerance / 2;
    const wasInThreshold = prevScrollPercent >= thresholdMin && prevScrollPercent <= thresholdMax;
    const isInThreshold = scrollPercent >= thresholdMin && scrollPercent <= thresholdMax;
    const hasEnteredThreshold = !wasInThreshold && isInThreshold && (!once || !thresholdFired);
    if (hasEnteredThreshold && once) {
      thresholdFired = true;
    }
    const callbackData = {
      scrollTop,
      scrollDirection,
      scrollPercent,
      directionChanged: scrollDirection !== prevScrollDirection,
      hasEnteredThreshold,
      isInThreshold
    };
    if (typeof callback === "function") {
      callback(callbackData);
    }
    if (scrollDirection !== prevScrollDirection) {
      if (scrollDirection === "down" && typeof onScrollDown === "function") {
        onScrollDown(callbackData);
      }
      if (scrollDirection === "up" && typeof onScrollUp === "function") {
        onScrollUp(callbackData);
      }
    }
    if (hasEnteredThreshold && typeof onEnterThreshold === "function") {
      onEnterThreshold(callbackData);
    }
    prevScrollTop = scrollTop;
    prevScrollPercent = scrollPercent;
    prevScrollDirection = scrollDirection;
  }
  return {
    destroy() {
      eventTarget.removeEventListener("scroll", throttledObserve);
      if (rafId) cancelAnimationFrame(rafId);
      cacheObserver.destroy();
    }
  };
}
export {
  scrollObserver
};
