// @splendidlabz/utils/dist/esm/lib/sse.js
import { parseJSON } from "./objects/json.js";
import { omitEmpty } from "./objects/omit-empty.js";
function createSSE({ event, data, id, retry }) {
  const eventString = event ? `event: ${event}
` : "";
  const idString = id ? `id: ${id}
` : "";
  const retryString = retry ? `retry: ${retry}
` : "";
  if (typeof data === "object") data = JSON.stringify(data);
  const dataString = data ? `data: ${data}
` : "";
  return `${eventString}${dataString}${idString}${retryString}
`;
}
function parseSSE(message) {
  const lines = message.split("\n");
  const result = {
    event: "message",
    // Default event type
    data: "",
    id: null,
    retry: null
  };
  for (const line of lines) {
    if (line.startsWith("data:")) {
      const data = line.slice(5).trim();
      result.data = parseJSON(data);
      continue;
    }
    if (line.startsWith("event:")) {
      result.event = line.slice(6).trim();
      continue;
    }
    if (line.startsWith("id:")) {
      result.id = line.slice(3).trim();
      continue;
    }
    if (line.startsWith("retry:")) {
      result.retry = parseInt(line.slice(6).trim(), 10);
      continue;
    }
  }
  const ret = omitEmpty(result);
  return ret.data ? ret : null;
}
export {
  createSSE,
  parseSSE
};
