// @splendidlabz/utils/dist/esm/node/file-cache.js
import fs from "fs/promises";
import { glob } from "glob";
import path from "node:path";
import { sort } from "../lib/arrays/index.js";
function resolveDirPath(dirPath) {
  return path.isAbsolute(dirPath) ? dirPath : path.join(process.cwd(), dirPath);
}
async function createCache(dir) {
  const absDirpath = resolveDirPath(dir);
  await fs.mkdir(absDirpath, { recursive: true });
  return {
    async cacheAt(file) {
      if (file) {
        const absFilepath = path.join(absDirpath, file);
        try {
          const stat = await fs.stat(absFilepath);
          return stat.mtimeMs;
        } catch (error) {
          return 0;
        }
      } else {
        return getLastModifiedTime(absDirpath + "/**/*");
      }
    },
    getAbsPath(file) {
      return path.join(absDirpath, file);
    },
    async hasFile(file) {
      const absFilepath = path.join(absDirpath, file);
      try {
        await fs.stat(absFilepath);
        return true;
      } catch (e) {
        return false;
      }
    },
    async loadFile(file) {
      try {
        const absFilepath = path.join(absDirpath, file);
        return fs.readFile(absFilepath);
      } catch (error) {
        console.warn(`File not found: ${file}`);
      }
    },
    async saveFile(file, data) {
      const absfilepath = path.join(absDirpath, file);
      await fs.writeFile(absfilepath, data);
    },
    async loadJSON(file) {
      const data = await this.loadFile(file);
      try {
        return JSON.parse(data.toString());
      } catch (err) {
        console.warn(`Error parsing JSON: ${file}`);
        return {};
      }
    },
    async saveJSON(file, data) {
      await this.saveFile(file, JSON.stringify(data));
    }
  };
}
async function fileCache({ dir, file }) {
  const absDirpath = resolveDirPath(dir);
  await fs.mkdir(absDirpath, { recursive: true });
  const cachePath = path.join(absDirpath, `${file}`);
  let modifiedAt = 0;
  try {
    const stat = await fs.stat(cachePath);
    modifiedAt = stat.mtimeMs;
  } catch (error) {
  }
  return {
    modifiedAt,
    async load() {
      return fs.readFile(cachePath);
    },
    async save(data) {
      await fs.writeFile(cachePath, data);
    },
    async loadJSON() {
      const data = await fs.readFile(cachePath);
      return JSON.parse(data.toString());
    },
    async saveJSON(data) {
      await fs.writeFile(cachePath, JSON.stringify(data));
    }
  };
}
async function getLastModifiedTime(globPath) {
  let files;
  if (typeof globPath === "string") {
    files = await glob(removeFirstSlash(globPath));
  }
  if (Array.isArray(globPath)) {
    const a = await Promise.all(
      globPath.map((path2) => glob(removeFirstSlash(path2)))
    );
    files = a.flat();
  }
  const stats = await Promise.all(files.map((file) => fs.stat(file)));
  const timestamps = stats.map((stat) => Number(stat.mtimeMs));
  const sorted = sort(timestamps, { props: "mtimeMs", reverse: true });
  return sorted[0] || 0;
}
async function getLatestModifiedTime(globPath) {
  console.warn(
    "`getLatestModifiedTime` has been renamed to `getLastModifiedTime` and will be deprecated in future versions. Please use `getLastModifiedTime` instead."
  );
  return getLastModifiedTime(globPath);
}
function removeFirstSlash(x) {
  return x.startsWith("/") ? x.slice(1) : x;
}
export {
  createCache,
  fileCache,
  getLastModifiedTime,
  getLatestModifiedTime,
  removeFirstSlash
};
