// @splendidlabz/utils/src/lib/functions/functional.js
/**
 * Creates a curried version of a function that can be called with partial arguments
 * @param {Function} fn - The function to curry
 * @return {Function} The curried function
 * @property {any} return - The result of calling the original function when all arguments are provided
 * @example
 * const add = (a, b, c) => a + b + c
 * const curriedAdd = curry(add)
 * curriedAdd(1)(2)(3) // 6
 * curriedAdd(1, 2)(3) // 6
 * curriedAdd(1)(2, 3) // 6
 */
export function curry(fn) {
  return function curried(...args) {
    if (args.length >= fn.length) {
      return fn.apply(this, args)
    } else {
      return function (...moreArgs) {
        return curried.apply(this, args.concat(moreArgs))
      }
    }
  }
}

/**
 * Composes functions from right to left (synchronous)
 * @param {...Function} fns - Functions to compose
 * @return {Function} The composed function
 * @property {any} return - The result of applying all functions in sequence
 * @example
 * const add1 = x => x + 1
 * const multiply2 = x => x * 2
 * const composed = compose(add1, multiply2)
 * composed(3) // 7 (3 * 2 + 1)
 */
export function compose(...fns) {
  return function (value) {
    return fns.reduceRight((acc, fn) => fn(acc), value)
  }
}

/**
 * Composes functions from right to left (asynchronous)
 * @param {...Function} fns - Functions to compose (can be sync or async)
 * @return {Function} The composed async function
 * @property {Promise<any>} return - A Promise that resolves to the result of applying all functions
 * @example
 * const add1 = x => x + 1
 * const multiplyAsync = async x => x * 2
 * await composeAsync(add1, multiplyAsync)(3) // 7 (3 * 2 + 1)
 */
export function composeAsync(...fns) {
  return async function (value) {
    return fns.reduceRight(async (acc, fn) => {
      const result = await acc
      return await fn(result)
    }, value)
  }
}

/**
 * Pipes functions from left to right (synchronous)
 * @param {...Function} fns - Functions to pipe
 * @return {Function} The piped function
 * @property {any} return - The result of applying all functions in sequence
 * @example
 * const add1 = x => x + 1
 * const multiply2 = x => x * 2
 * const piped = pipe(add1, multiply2)
 * piped(3) // 8 (3 + 1) * 2
 */
export function pipe(...fns) {
  return function (value) {
    return fns.reduce((acc, fn) => fn(acc), value)
  }
}

/**
 * Pipes functions from left to right (asynchronous)
 * @param {...Function} fns - Functions to pipe (can be sync or async)
 * @return {Function} The piped async function
 * @property {Promise<any>} return - A Promise that resolves to the result of applying all functions
 * @example
 * const add1 = x => x + 1
 * const multiplyAsync = async x => x * 2
 * await pipeAsync(add1, multiplyAsync)(3) // 8 ((3 + 1) * 2)
 */
export function pipeAsync(...fns) {
  return async function (value) {
    return fns.reduce(async (acc, fn) => {
      const result = await acc
      return await fn(result)
    }, value)
  }
}

/**
 * Calls a function n times with the current index and returns an array of results
 * @param {Function} fn - Function to call (receives index as argument)
 * @param {number} n - Number of times to call the function
 * @return {any[]} Array of results from calling the function
 * @property {any[]} return - Array containing the results of each function call
 * @example
 * times(i => i * 2, 3) // [0, 2, 4]
 * times(() => Math.random(), 2) // [0.123, 0.456] (random values)
 */
export function times(fn, n) {
  const result = []
  for (let i = 0; i < n; i++) {
    result.push(fn(i))
  }
  return result
}
