// @splendidlabz/utils/dist/cjs/lib/strings/convert-case/convert-case.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var convert_case_exports = {};
__export(convert_case_exports, {
  toCamel: () => toCamel,
  toKebab: () => toKebab,
  toLower: () => toLower,
  toPascal: () => toPascal,
  toSentence: () => toSentence,
  toSlug: () => toSlug,
  toTitle: () => toTitle,
  toUpper: () => toUpper
});
module.exports = __toCommonJS(convert_case_exports);
function toKebab(string) {
  if (!string) return "";
  return string.split("").map((letter, index) => {
    const previousLetter = string[index - 1] || "";
    const currentLetter = letter;
    if (isDigit(currentLetter) && !isDigit(previousLetter)) {
      return `-${currentLetter}`;
    }
    if (!isCaps(currentLetter)) return currentLetter;
    if (previousLetter === "") {
      return `${currentLetter.toLowerCase()}`;
    }
    if (isCaps(previousLetter)) {
      return `${currentLetter.toLowerCase()}`;
    }
    return `-${currentLetter.toLowerCase()}`;
  }).join("").trim().replace(/[-_\s]+/g, "-");
}
function toCamel(string) {
  return toKebab(string).split("-").map((word, index) => {
    if (index === 0) return word;
    return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
  }).join("");
}
function toPascal(string) {
  const camel = toCamel(string);
  return camel.slice(0, 1).toUpperCase() + camel.slice(1);
}
function toTitle(string) {
  return toKebab(string).split("-").map((word) => {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
  }).join(" ");
}
function toSlug(string) {
  return toKebab(string);
}
function toSentence(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.slice(0, 1).toUpperCase() + interim.slice(1);
}
function toUpper(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.toUpperCase();
}
function toLower(string) {
  const interim = toKebab(string).replace(/-/g, " ");
  return interim.toLowerCase();
}
function isCaps(char) {
  return /\p{Lu}/u.test(char);
}
function isDigit(char) {
  return /[0-9]/.test(char);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toCamel,
  toKebab,
  toLower,
  toPascal,
  toSentence,
  toSlug,
  toTitle,
  toUpper
});
