// @splendidlabz/utils/dist/cjs/lib/functions/functional.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var functional_exports = {};
__export(functional_exports, {
  compose: () => compose,
  composeAsync: () => composeAsync,
  curry: () => curry,
  pipe: () => pipe,
  pipeAsync: () => pipeAsync,
  times: () => times
});
module.exports = __toCommonJS(functional_exports);
function curry(fn) {
  return function curried(...args) {
    if (args.length >= fn.length) {
      return fn.apply(this, args);
    } else {
      return function(...moreArgs) {
        return curried.apply(this, args.concat(moreArgs));
      };
    }
  };
}
function compose(...fns) {
  return function(value) {
    return fns.reduceRight((acc, fn) => fn(acc), value);
  };
}
function composeAsync(...fns) {
  return async function(value) {
    return fns.reduceRight(async (acc, fn) => {
      const result = await acc;
      return await fn(result);
    }, value);
  };
}
function pipe(...fns) {
  return function(value) {
    return fns.reduce((acc, fn) => fn(acc), value);
  };
}
function pipeAsync(...fns) {
  return async function(value) {
    return fns.reduce(async (acc, fn) => {
      const result = await acc;
      return await fn(result);
    }, value);
  };
}
function times(fn, n) {
  const result = [];
  for (let i = 0; i < n; i++) {
    result.push(fn(i));
  }
  return result;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  compose,
  composeAsync,
  curry,
  pipe,
  pipeAsync,
  times
});
