// @splendidlabz/utils/src/lib/sse.js
// Server Sent Events (SSE) 
import { parseJSON } from './objects/json.js'
import { omitEmpty } from './objects/omit-empty.js'

export function createSSE({ event, data, id, retry }) {
  const eventString = event ? `event: ${event}\n` : ''
  const idString = id ? `id: ${id}\n` : ''
  const retryString = retry ? `retry: ${retry}\n` : ''
  
  if (typeof data === 'object') data = JSON.stringify(data)
  const dataString = data ? `data: ${data}\n` : ''

  return `${eventString}${dataString}${idString}${retryString}\n`
}

/**
 * Parses a single SSE message into a structured object
 * @param {string} message - The raw SSE message
 * @returns {Object|null} The parsed message or null if invalid
 * @property {string} [event] - The event type if specified
 * @property {string|Object} data - The message data (parsed as JSON if possible)
 * @property {string} [id] - The message ID if specified
 * @property {number} [retry] - The retry interval if specified
 */
export function parseSSE(message) {
  const lines = message.split('\n')
  const result = {
    event: 'message', // Default event type
    data: '',
    id: null,
    retry: null
  }

  for (const line of lines) {
    if (line.startsWith('data:')) {
      const data = line.slice(5).trim()
      result.data = parseJSON(data)
      continue
    } 
    
    if (line.startsWith('event:')) {
      result.event = line.slice(6).trim()
      continue
    } 
    
    if (line.startsWith('id:')) {
      result.id = line.slice(3).trim()
      continue
    } 

    if (line.startsWith('retry:')) {
      result.retry = parseInt(line.slice(6).trim(), 10)
      continue
    }
  }


  const ret = omitEmpty(result)
  return ret.data ? ret : null
}
