// @splendidlabz/utils/src/dom/bounding-box.js
/**
 * Advanced boundingClientRect that contains information vertical and horizontal centers.
 * @param {HTMLElement} element - The element to get the bounding box for
 * @param {Object} [options] - Options for the bounding box calculation
 * @param {boolean} [options.containsPadding=true] - Whether to include padding in the calculations
 * @returns {Object} The bounding box object with additional center coordinates
 * @property {number} x - The x coordinate
 * @property {number} y - The y coordinate
 * @property {number} width - The width
 * @property {number} height - The height
 * @property {number} top - The top coordinate
 * @property {number} right - The right coordinate
 * @property {number} bottom - The bottom coordinate
 * @property {number} left - The left coordinate
 * @property {number} xCenter - The horizontal center coordinate
 * @property {number} yCenter - The vertical center coordinate
 */
export function boundingBox(element, { containsPadding = true } = {}) {
  const box = element.getBoundingClientRect()

  // Spread original properties into the return object
  const ret = {}
  for (const prop in box) ret[prop] = box[prop]

  // Gets boundingClientRect value without padding
  if (!containsPadding) {
    const style = getComputedStyle(element)
    // Parse padding values immediately to numbers
    const paddingTop = parseFloat(style.paddingTop)
    const paddingRight = parseFloat(style.paddingRight)
    const paddingBottom = parseFloat(style.paddingBottom)
    const paddingLeft = parseFloat(style.paddingLeft)

    ret.x = ret.x + paddingLeft
    ret.y = ret.y + paddingTop
    ret.width = ret.width - paddingLeft - paddingRight
    ret.height = ret.height - paddingTop - paddingBottom
    ret.top = ret.top + paddingTop
    ret.bottom = ret.bottom - paddingBottom
    ret.right = ret.right - paddingRight
    ret.left = ret.left + paddingLeft
  }

  ret.xCenter = (ret.left + ret.right) / 2
  ret.yCenter = (ret.top + ret.bottom) / 2
  return ret
}

/**
 * Gets the bounding box of an element relative to its ancestor
 * @param {Object} params - Parameters for the calculation
 * @param {HTMLElement} params.element - The element to get the bounding box for
 * @param {HTMLElement} [params.ancestorElement] - The ancestor element to calculate position relative to. Defaults to parent element
 * @returns {Object} The relative bounding box with center coordinates
 * @property {number} top - Distance from ancestor's top
 * @property {number} right - Distance from ancestor's right
 * @property {number} bottom - Distance from ancestor's bottom
 * @property {number} left - Distance from ancestor's left
 * @property {number} width - The width
 * @property {number} height - The height
 * @property {number} xCenter - The horizontal center coordinate relative to ancestor
 * @property {number} yCenter - The vertical center coordinate relative to ancestor
 */
export function boundingBoxRelativeToAncestor({ element, ancestorElement }) {
  ancestorElement = ancestorElement || element.parentElement

  const box = boundingBox(element)
  const ancestorBox = boundingBox(ancestorElement, {
    containsPadding: false,
  })

  return {
    top: box.top - ancestorBox.top,
    right: box.right - ancestorBox.right,
    bottom: box.bottom - ancestorBox.bottom,
    left: box.left - ancestorBox.left,
    width: box.width,
    height: box.height,
    xCenter: (box.left + box.right) / 2 - ancestorBox.left,
    yCenter: (box.top + box.bottom) / 2 - ancestorBox.top,
  }
}
