// @splendidlabz/utils/src/dom/focusable.js
/**
 * @typedef {Object} Focusables
 * @property {number} length - Number of keyboard focusable elements
 * @property {Array} all - All focusable elements that are visible and enabled
 * @property {Array} keyboard - Elements that can be focused via keyboard (tabIndex > -1)
 * @property {Array} keyboardOnly - Alias for keyboard focusable elements
 * @property {HTMLElement|null} first - First keyboard focusable element
 * @property {HTMLElement|null} last - Last keyboard focusable element
 * @property {function(number): HTMLElement|null} next - Get next focusable element after the given index
 * @property {function(number): HTMLElement|null} prev - Get previous focusable element before the given index
 */

/**
 * Gets all focusable elements within a container and provides methods to navigate between them.
 * This is useful for implementing keyboard navigation, focus trapping, and accessibility features.
 * 
 * @param {HTMLElement} [container=document.body] - The container element to search within. Only elements inside this container will be included.
 * @returns {Focusables} Object containing focusable elements and navigation methods
 * @example
 * // Get focusable elements in a modal
 * const modal = document.querySelector('.modal')
 * const focusable = getFocusableElements(modal)
 * 
 * // Get all focusable elements in document
 * const allFocusable = getFocusableElements()
 * 
 * // Focus the first element when modal opens
 * focusable.first?.focus()
 * 
 * // Handle tab key to trap focus
 * modal.addEventListener('keydown', e => {
 *   if (e.key === 'Tab') {
 *     e.preventDefault()
 *     const currentIndex = focusable.keyboard.indexOf(document.activeElement)
 *     const nextElement = e.shiftKey 
 *       ? focusable.prev(currentIndex) 
 *       : focusable.next(currentIndex)
 *     nextElement?.focus()
 *   }
 * })
 */
export function getFocusableElements(container = document.body) {
  return {
    /**
     * Gets the number of keyboard focusable elements.
     * This is useful for checking if there are any focusable elements
     * before attempting to focus them.
     * 
     * @returns {number} The number of keyboard focusable elements
     */
    get length() {
      return this.keyboard.length
    },

    /**
     * Gets all focusable elements within the container that are visible and enabled.
     * This includes elements that can be focused via mouse or programmatically,
     * but may not be keyboard focusable.
     * 
     * Elements are considered focusable if they:
     * - Are not disabled
     * - Are not hidden
     * - Are not display:none
     * 
     * @returns {HTMLElement[]} Array of focusable elements that are visible and enabled
     */
    get all() {
      const elements = Array.from(
        container.querySelectorAll(
          `a, 
            button, 
            input, 
            textarea, 
            select, 
            details, 
            iframe, 
            embed, 
            object, 
            summary, 
            dialog, 
            audio[controls], 
            video[controls], 
            [contenteditable], 
            [tabindex]
          `,
        ),
      )
      
      // @ts-ignore - we check instanceof HTMLElement at runtime
      return elements.filter(el => {
        if (el.hasAttribute('disabled')) return false
        if (el.hasAttribute('hidden')) return false
        if (window.getComputedStyle(el).display === 'none') return false
        return true
      })
    },

    /**
     * Gets elements that can be focused via keyboard (tabIndex > -1).
     * These are the elements that will be focused when using the Tab key.
     * 
     * @returns {HTMLElement[]} Array of keyboard focusable elements
     */
    get keyboard() {
      return this.all.filter(el => el instanceof HTMLElement && el.tabIndex > -1)
    },

    /**
     * Alias for keyboard focusable elements.
     * This is provided for semantic clarity in certain contexts.
     * 
     * @returns {HTMLElement[]} Array of keyboard focusable elements
     */
    get keyboardOnly() {
      return this.keyboard
    },

    /**
     * Gets the first keyboard focusable element.
     * This is useful for setting initial focus when opening a modal or dialog.
     * 
     * @returns {HTMLElement|null} The first focusable element or null if none exist
     */
    get first() {
      return this.keyboard[0] || null
    },

    /**
     * Gets the last keyboard focusable element.
     * This is useful for focus trapping when using Shift+Tab.
     * 
     * @returns {HTMLElement|null} The last focusable element or null if none exist
     */
    get last() {
      return this.keyboard[this.length - 1] || null
    },

    /**
     * Gets the next focusable element after the given index.
     * This is useful for implementing custom tab order or focus management.
     * 
     * @param {number} index - The current element's index in the keyboard focusable array
     * @returns {HTMLElement|null} The next focusable element or null if none exists
     */
    next(index) {
      return this.keyboard[index + 1] || null
    },

    /**
     * Gets the previous focusable element before the given index.
     * This is useful for implementing custom tab order or focus management.
     * 
     * @param {number} index - The current element's index in the keyboard focusable array
     * @returns {HTMLElement|null} The previous focusable element or null if none exists
     */
    prev(index) {
      return this.keyboard[index - 1] || null
    },
  }
}

// Alias of getFocusableElements
export function Focusable(container = document.body) {
  return getFocusableElements(container)
}
