// @splendidlabz/utils/src/dom/font-size.js
import { splitUnit } from '../lib/numbers/index.js'

export function getUnit(value) {
  const [, unit] = splitUnit(value)
  return unit
}

export function rem(multiple = 1) {
  const value = parseFloat(getComputedStyle(document.body)['font-size'])
  return Math.round(value * multiple)
}

export function em(element = document.body, multiple = 1) {
  const value = parseFloat(getComputedStyle(element)['font-size'])
  return Math.round(value * multiple)
}

export function rlh(multiple = 1) {
  const lineHeight = parseFloat(getComputedStyle(document.body)['line-height'])
  return Math.round(lineHeight * multiple)
}

export function lh(element = document.body, multiple = 1) {
  const lineHeight = getComputedStyle(element)['line-height']
  const value = parseFloat(lineHeight)
  return Math.round(value * multiple)
}

export function toPx(value, element = null) {
  const [numeric, unit] = splitUnit(value)
  let finalValue = 0

  if (unit === 'rem') finalValue = rem(numeric)
  if (unit === 'em') finalValue = em(element, numeric)
  if (unit === 'lh') finalValue = lh(element, numeric)
  if (unit === 'px') finalValue = numeric
  if (!unit) finalValue = numeric // Default to px offsets

  return finalValue
}
