// @splendidlabz/utils/src/dom/get-element.js
// Returns the target HTMLElement.
// If the selector is already an HTMLElement, it will return the selector itself.
// Otherwise, it will return the first element that matches the selector.

/**
 * List of Astro-specific node names that need special handling
 * @type {string[]}
 */
const astroNodes = ['ASTRO-SLOT', 'ASTRO-ISLAND']

/**
 * Gets the type of a DOM node
 * @param {Node|NodeList|Array} node - The node to check
 * @returns {'element'|'nodelist'|'array'|undefined} The type of the node
 */
export function getNodeType(node) {
  if (node instanceof Element) return 'element'
  if (node instanceof NodeList) return 'nodelist'
  if (Array.isArray(node)) return 'array'
}

/**
 * Gets an element from a selector or returns the element if it's already an Element
 * @param {string|Element} selector - CSS selector or Element
 * @returns {Element|undefined} The matching element or undefined if not found
 * 
 * @example
 * // Get element by selector
 * const button = getElement('.button')
 * 
 * // Pass through Element
 * const element = document.querySelector('.button')
 * const sameElement = getElement(element) // Returns element directly
 */
export function getElement(selector) {
  if (!selector) return
  if (selector instanceof Element) return selector
  return document.querySelector(selector)
}

/**
 * Gets all child elements of a node, handling Astro slots
 * @param {Element} node - The parent node
 * @returns {Element[]} Array of child elements
 * 
 * @example
 * // Get children of a div
 * const children = getChildrenElements(div)
 * 
 * // Works with Astro slots
 * const slotChildren = getChildrenElements(astroSlot)
 */
export function getChildrenElements(node) {
  let children = node.children

  // Compensate for Astro Slots
  if (children[0]?.nodeName === 'ASTRO-SLOT') children = children[0].children
  return Array.from(children)
}

/**
 * Gets the parent element, skipping Astro-specific nodes
 * @param {Element} element - The element to get the parent of
 * @returns {Element|null} The parent element, or null if no parent exists
 * 
 * @example
 * // Get parent, skipping Astro slots
 * const parent = getParentElement(element)
 */
export function getParentElement(element) {
  const parent = element.parentElement
  if (astroNodes.includes(parent.nodeName)) return getParentElement(parent)
  return parent
}

/**
 * Gets all sibling elements of an element (excluding the element itself)
 * @param {Element} element - The element to get siblings of
 * @returns {Element[]} Array of sibling elements
 * 
 * @example
 * // Get all siblings of an element
 * const siblings = getSiblingElements(element)
 */
export function getSiblingElements(element) {
  return Array.from(element.parentElement.children).filter(
    child => child !== element
  )
}

/**
 * Gets the index of an element among its siblings
 * @param {Element} element - The element to get the index of
 * @returns {number} The index of the element (0-based)
 * 
 * @example
 * // Get position of element among siblings
 * const index = getSelfIndex(element)
 */
export function getSelfIndex(element) {
  return Array.from(element.parentElement.children).indexOf(element)
}

/**
 * Checks if an element is an ancestor of another element
 * @param {Element} ancestor - The potential ancestor element
 * @param {Element} element - The element to check
 * @param {Object} [options] - Options for the search
 * @param {number} [options.searchLevels=0] - Number of levels to search up, 0 for unlimited
 * @returns {boolean} True if ancestor is an ancestor of element
 * 
 * @example
 * // Check if div is an ancestor of button
 * const isParent = isAncestor(div, button)
 * 
 * // Check if div is an ancestor within 2 levels
 * const isCloseAncestor = isAncestor(div, button, { searchLevels: 2 })
 */
export function isAncestor(ancestor, element, { searchLevels = 0 } = {}) {
  if (ancestor === element) return true

  let parent = element.parentElement
  // Infinite Search
  if (searchLevels === 0) {
    while (parent) {
      if (ancestor === parent) return true
      parent = parent.parentElement
    }
  } else {
    // Search X levels upwards
    for (let i = 0; i < searchLevels; i++) {
      if (ancestor === parent) return true
      parent = parent.parentElement
    }
  }

  return false
}

/**
 * Gets ancestor elements with their siblings, optionally limiting the search depth.
 * Useful for finding the nearest ancestor that has siblings.
 * 
 * @param {Element} element - The element to start searching from
 * @param {Object} [options] - Options for the search
 * @param {number|boolean} [options.limit=false] - Number of levels to search up, or false for unlimited
 * @returns {Element[]|null} Array of sibling elements, or null if none found
 * 
 * @example
 * // Get siblings of nearest ancestor that has siblings
 * const ancestorSiblings = getAncestorWithSiblings(element)
 * 
 * // Search up to 2 levels
 * const limitedSiblings = getAncestorWithSiblings(element, { limit: 2 })
 */
export function getAncestorWithSiblings(element, { limit = false } = {}) {
  const ancestor = element.parentElement
  const siblings = getSiblingElements(ancestor)

  // Found
  if (siblings.length) return siblings

  // No more ancestors
  if (!ancestor) return null

  // Unlimited upwards search
  if (limit === false) return getAncestorWithSiblings(ancestor)

  // Limited search
  if (typeof limit === 'number' && limit > 0) {
    return getAncestorWithSiblings(ancestor, { limit: limit - 1 })
  }

  // Limit reached, end search
  if (limit === 0) return null
}
